// dotdog/dist/cli.js
#!/usr/bin/env node
import { createRequire } from "node:module";
var __require = /* @__PURE__ */ createRequire(import.meta.url);

// ../../node_modules/.bun/commander@15.0.0/node_modules/commander/lib/error.js
class CommanderError extends Error {
  constructor(exitCode, code, message) {
    super(message);
    Error.captureStackTrace(this, this.constructor);
    this.name = this.constructor.name;
    this.code = code;
    this.exitCode = exitCode;
    this.nestedError = undefined;
  }
}

class InvalidArgumentError extends CommanderError {
  constructor(message) {
    super(1, "commander.invalidArgument", message);
    Error.captureStackTrace(this, this.constructor);
    this.name = this.constructor.name;
  }
}

// ../../node_modules/.bun/commander@15.0.0/node_modules/commander/lib/argument.js
class Argument {
  constructor(name, description) {
    this.description = description || "";
    this.variadic = false;
    this.parseArg = undefined;
    this.defaultValue = undefined;
    this.defaultValueDescription = undefined;
    this.argChoices = undefined;
    switch (name[0]) {
      case "<":
        this.required = true;
        this._name = name.slice(1, -1);
        break;
      case "[":
        this.required = false;
        this._name = name.slice(1, -1);
        break;
      default:
        this.required = true;
        this._name = name;
        break;
    }
    if (this._name.endsWith("...")) {
      this.variadic = true;
      this._name = this._name.slice(0, -3);
    }
  }
  name() {
    return this._name;
  }
  _collectValue(value, previous) {
    if (previous === this.defaultValue || !Array.isArray(previous)) {
      return [value];
    }
    previous.push(value);
    return previous;
  }
  default(value, description) {
    this.defaultValue = value;
    this.defaultValueDescription = description;
    return this;
  }
  argParser(fn) {
    this.parseArg = fn;
    return this;
  }
  choices(values) {
    this.argChoices = values.slice();
    this.parseArg = (arg, previous) => {
      if (!this.argChoices.includes(arg)) {
        throw new InvalidArgumentError(`Allowed choices are ${this.argChoices.join(", ")}.`);
      }
      if (this.variadic) {
        return this._collectValue(arg, previous);
      }
      return arg;
    };
    return this;
  }
  argRequired() {
    this.required = true;
    return this;
  }
  argOptional() {
    this.required = false;
    return this;
  }
}
function humanReadableArgName(arg) {
  const nameOutput = arg.name() + (arg.variadic === true ? "..." : "");
  return arg.required ? "<" + nameOutput + ">" : "[" + nameOutput + "]";
}

// ../../node_modules/.bun/commander@15.0.0/node_modules/commander/lib/command.js
import { EventEmitter } from "node:events";
import childProcess from "node:child_process";
import path from "node:path";
import fs from "node:fs";
import process2 from "node:process";
import { stripVTControlCharacters as stripVTControlCharacters2 } from "node:util";

// ../../node_modules/.bun/commander@15.0.0/node_modules/commander/lib/help.js
import { stripVTControlCharacters } from "node:util";

class Help {
  constructor() {
    this.helpWidth = undefined;
    this.minWidthToWrap = 40;
    this.sortSubcommands = false;
    this.sortOptions = false;
    this.showGlobalOptions = false;
  }
  prepareContext(contextOptions) {
    this.helpWidth = this.helpWidth ?? contextOptions.helpWidth ?? 80;
  }
  visibleCommands(cmd) {
    const visibleCommands = cmd.commands.filter((cmd2) => !cmd2._hidden);
    const helpCommand = cmd._getHelpCommand();
    if (helpCommand && !helpCommand._hidden) {
      visibleCommands.push(helpCommand);
    }
    if (this.sortSubcommands) {
      visibleCommands.sort((a, b) => {
        return a.name().localeCompare(b.name());
      });
    }
    return visibleCommands;
  }
  compareOptions(a, b) {
    const getSortKey = (option) => {
      return option.short ? option.short.replace(/^-/, "") : option.long.replace(/^--/, "");
    };
    return getSortKey(a).localeCompare(getSortKey(b));
  }
  visibleOptions(cmd) {
    const visibleOptions = cmd.options.filter((option) => !option.hidden);
    const helpOption = cmd._getHelpOption();
    if (helpOption && !helpOption.hidden) {
      const removeShort = helpOption.short && cmd._findOption(helpOption.short);
      const removeLong = helpOption.long && cmd._findOption(helpOption.long);
      if (!removeShort && !removeLong) {
        visibleOptions.push(helpOption);
      } else if (helpOption.long && !removeLong) {
        visibleOptions.push(cmd.createOption(helpOption.long, helpOption.description));
      } else if (helpOption.short && !removeShort) {
        visibleOptions.push(cmd.createOption(helpOption.short, helpOption.description));
      }
    }
    if (this.sortOptions) {
      visibleOptions.sort(this.compareOptions);
    }
    return visibleOptions;
  }
  visibleGlobalOptions(cmd) {
    if (!this.showGlobalOptions)
      return [];
    const globalOptions = [];
    for (let ancestorCmd = cmd.parent;ancestorCmd; ancestorCmd = ancestorCmd.parent) {
      const visibleOptions = ancestorCmd.options.filter((option) => !option.hidden);
      globalOptions.push(...visibleOptions);
    }
    if (this.sortOptions) {
      globalOptions.sort(this.compareOptions);
    }
    return globalOptions;
  }
  visibleArguments(cmd) {
    if (cmd._argsDescription) {
      cmd.registeredArguments.forEach((argument) => {
        argument.description = argument.description || cmd._argsDescription[argument.name()] || "";
      });
    }
    if (cmd.registeredArguments.find((argument) => argument.description)) {
      return cmd.registeredArguments;
    }
    return [];
  }
  subcommandTerm(cmd) {
    const args = cmd.registeredArguments.map((arg) => humanReadableArgName(arg)).join(" ");
    return cmd._name + (cmd._aliases[0] ? "|" + cmd._aliases[0] : "") + (cmd.options.length ? " [options]" : "") + (args ? " " + args : "");
  }
  optionTerm(option) {
    return option.flags;
  }
  argumentTerm(argument) {
    return argument.name();
  }
  longestSubcommandTermLength(cmd, helper) {
    return helper.visibleCommands(cmd).reduce((max, command) => {
      return Math.max(max, this.displayWidth(helper.styleSubcommandTerm(helper.subcommandTerm(command))));
    }, 0);
  }
  longestOptionTermLength(cmd, helper) {
    return helper.visibleOptions(cmd).reduce((max, option) => {
      return Math.max(max, this.displayWidth(helper.styleOptionTerm(helper.optionTerm(option))));
    }, 0);
  }
  longestGlobalOptionTermLength(cmd, helper) {
    return helper.visibleGlobalOptions(cmd).reduce((max, option) => {
      return Math.max(max, this.displayWidth(helper.styleOptionTerm(helper.optionTerm(option))));
    }, 0);
  }
  longestArgumentTermLength(cmd, helper) {
    return helper.visibleArguments(cmd).reduce((max, argument) => {
      return Math.max(max, this.displayWidth(helper.styleArgumentTerm(helper.argumentTerm(argument))));
    }, 0);
  }
  commandUsage(cmd) {
    let cmdName = cmd._name;
    if (cmd._aliases[0]) {
      cmdName = cmdName + "|" + cmd._aliases[0];
    }
    let ancestorCmdNames = "";
    for (let ancestorCmd = cmd.parent;ancestorCmd; ancestorCmd = ancestorCmd.parent) {
      ancestorCmdNames = ancestorCmd.name() + " " + ancestorCmdNames;
    }
    return ancestorCmdNames + cmdName + " " + cmd.usage();
  }
  commandDescription(cmd) {
    return cmd.description();
  }
  subcommandDescription(cmd) {
    return cmd.summary() || cmd.description();
  }
  optionDescription(option) {
    const extraInfo = [];
    if (option.argChoices) {
      extraInfo.push(`choices: ${option.argChoices.map((choice) => JSON.stringify(choice)).join(", ")}`);
    }
    if (option.defaultValue !== undefined) {
      const showDefault = option.required || option.optional || option.isBoolean() && typeof option.defaultValue === "boolean";
      if (showDefault) {
        extraInfo.push(`default: ${option.defaultValueDescription || JSON.stringify(option.defaultValue)}`);
      }
    }
    if (option.presetArg !== undefined && option.optional) {
      extraInfo.push(`preset: ${JSON.stringify(option.presetArg)}`);
    }
    if (option.envVar !== undefined) {
      extraInfo.push(`env: ${option.envVar}`);
    }
    if (extraInfo.length > 0) {
      const extraDescription = `(${extraInfo.join(", ")})`;
      if (option.description) {
        return `${option.description} ${extraDescription}`;
      }
      return extraDescription;
    }
    return option.description;
  }
  argumentDescription(argument) {
    const extraInfo = [];
    if (argument.argChoices) {
      extraInfo.push(`choices: ${argument.argChoices.map((choice) => JSON.stringify(choice)).join(", ")}`);
    }
    if (argument.defaultValue !== undefined) {
      extraInfo.push(`default: ${argument.defaultValueDescription || JSON.stringify(argument.defaultValue)}`);
    }
    if (extraInfo.length > 0) {
      const extraDescription = `(${extraInfo.join(", ")})`;
      if (argument.description) {
        return `${argument.description} ${extraDescription}`;
      }
      return extraDescription;
    }
    return argument.description;
  }
  formatItemList(heading, items, helper) {
    if (items.length === 0)
      return [];
    return [helper.styleTitle(heading), ...items, ""];
  }
  groupItems(unsortedItems, visibleItems, getGroup) {
    const result = new Map;
    unsortedItems.forEach((item) => {
      const group = getGroup(item);
      if (!result.has(group))
        result.set(group, []);
    });
    visibleItems.forEach((item) => {
      const group = getGroup(item);
      if (!result.has(group)) {
        result.set(group, []);
      }
      result.get(group).push(item);
    });
    return result;
  }
  formatHelp(cmd, helper) {
    const termWidth = helper.padWidth(cmd, helper);
    const helpWidth = helper.helpWidth ?? 80;
    function callFormatItem(term, description) {
      return helper.formatItem(term, termWidth, description, helper);
    }
    let output = [
      `${helper.styleTitle("Usage:")} ${helper.styleUsage(helper.commandUsage(cmd))}`,
      ""
    ];
    const commandDescription = helper.commandDescription(cmd);
    if (commandDescription.length > 0) {
      output = output.concat([
        helper.boxWrap(helper.styleCommandDescription(commandDescription), helpWidth),
        ""
      ]);
    }
    const argumentList = helper.visibleArguments(cmd).map((argument) => {
      return callFormatItem(helper.styleArgumentTerm(helper.argumentTerm(argument)), helper.styleArgumentDescription(helper.argumentDescription(argument)));
    });
    output = output.concat(this.formatItemList("Arguments:", argumentList, helper));
    const optionGroups = this.groupItems(cmd.options, helper.visibleOptions(cmd), (option) => option.helpGroupHeading ?? "Options:");
    optionGroups.forEach((options, group) => {
      const optionList = options.map((option) => {
        return callFormatItem(helper.styleOptionTerm(helper.optionTerm(option)), helper.styleOptionDescription(helper.optionDescription(option)));
      });
      output = output.concat(this.formatItemList(group, optionList, helper));
    });
    if (helper.showGlobalOptions) {
      const globalOptionList = helper.visibleGlobalOptions(cmd).map((option) => {
        return callFormatItem(helper.styleOptionTerm(helper.optionTerm(option)), helper.styleOptionDescription(helper.optionDescription(option)));
      });
      output = output.concat(this.formatItemList("Global Options:", globalOptionList, helper));
    }
    const commandGroups = this.groupItems(cmd.commands, helper.visibleCommands(cmd), (sub) => sub.helpGroup() || "Commands:");
    commandGroups.forEach((commands, group) => {
      const commandList = commands.map((sub) => {
        return callFormatItem(helper.styleSubcommandTerm(helper.subcommandTerm(sub)), helper.styleSubcommandDescription(helper.subcommandDescription(sub)));
      });
      output = output.concat(this.formatItemList(group, commandList, helper));
    });
    return output.join(`
`);
  }
  displayWidth(str) {
    return stripVTControlCharacters(str).length;
  }
  styleTitle(str) {
    return str;
  }
  styleUsage(str) {
    return str.split(" ").map((word) => {
      if (word === "[options]")
        return this.styleOptionText(word);
      if (word === "[command]")
        return this.styleSubcommandText(word);
      if (word[0] === "[" || word[0] === "<")
        return this.styleArgumentText(word);
      return this.styleCommandText(word);
    }).join(" ");
  }
  styleCommandDescription(str) {
    return this.styleDescriptionText(str);
  }
  styleOptionDescription(str) {
    return this.styleDescriptionText(str);
  }
  styleSubcommandDescription(str) {
    return this.styleDescriptionText(str);
  }
  styleArgumentDescription(str) {
    return this.styleDescriptionText(str);
  }
  styleDescriptionText(str) {
    return str;
  }
  styleOptionTerm(str) {
    return this.styleOptionText(str);
  }
  styleSubcommandTerm(str) {
    return str.split(" ").map((word) => {
      if (word === "[options]")
        return this.styleOptionText(word);
      if (word[0] === "[" || word[0] === "<")
        return this.styleArgumentText(word);
      return this.styleSubcommandText(word);
    }).join(" ");
  }
  styleArgumentTerm(str) {
    return this.styleArgumentText(str);
  }
  styleOptionText(str) {
    return str;
  }
  styleArgumentText(str) {
    return str;
  }
  styleSubcommandText(str) {
    return str;
  }
  styleCommandText(str) {
    return str;
  }
  padWidth(cmd, helper) {
    return Math.max(helper.longestOptionTermLength(cmd, helper), helper.longestGlobalOptionTermLength(cmd, helper), helper.longestSubcommandTermLength(cmd, helper), helper.longestArgumentTermLength(cmd, helper));
  }
  preformatted(str) {
    return /\n[^\S\r\n]/.test(str);
  }
  formatItem(term, termWidth, description, helper) {
    const itemIndent = 2;
    const itemIndentStr = " ".repeat(itemIndent);
    if (!description)
      return itemIndentStr + term;
    const paddedTerm = term.padEnd(termWidth + term.length - helper.displayWidth(term));
    const spacerWidth = 2;
    const helpWidth = this.helpWidth ?? 80;
    const remainingWidth = helpWidth - termWidth - spacerWidth - itemIndent;
    let formattedDescription;
    if (remainingWidth < this.minWidthToWrap || helper.preformatted(description)) {
      formattedDescription = description;
    } else {
      const wrappedDescription = helper.boxWrap(description, remainingWidth);
      formattedDescription = wrappedDescription.replace(/\n/g, `
` + " ".repeat(termWidth + spacerWidth));
    }
    return itemIndentStr + paddedTerm + " ".repeat(spacerWidth) + formattedDescription.replace(/\n/g, `
${itemIndentStr}`);
  }
  boxWrap(str, width) {
    if (width < this.minWidthToWrap)
      return str;
    const rawLines = str.split(/\r\n|\n/);
    const chunkPattern = /[\s]*[^\s]+/g;
    const wrappedLines = [];
    rawLines.forEach((line) => {
      const chunks = line.match(chunkPattern);
      if (chunks === null) {
        wrappedLines.push("");
        return;
      }
      let sumChunks = [chunks.shift()];
      let sumWidth = this.displayWidth(sumChunks[0]);
      chunks.forEach((chunk) => {
        const visibleWidth = this.displayWidth(chunk);
        if (sumWidth + visibleWidth <= width) {
          sumChunks.push(chunk);
          sumWidth += visibleWidth;
          return;
        }
        wrappedLines.push(sumChunks.join(""));
        const nextChunk = chunk.trimStart();
        sumChunks = [nextChunk];
        sumWidth = this.displayWidth(nextChunk);
      });
      wrappedLines.push(sumChunks.join(""));
    });
    return wrappedLines.join(`
`);
  }
}

// ../../node_modules/.bun/commander@15.0.0/node_modules/commander/lib/option.js
class Option {
  constructor(flags, description) {
    this.flags = flags;
    this.description = description || "";
    this.required = flags.includes("<");
    this.optional = flags.includes("[");
    this.variadic = /\w\.\.\.[>\]]$/.test(flags);
    this.mandatory = false;
    const optionFlags = splitOptionFlags(flags);
    this.short = optionFlags.shortFlag;
    this.long = optionFlags.longFlag;
    this.negate = false;
    if (this.long) {
      this.negate = this.long.startsWith("--no-");
    }
    this.defaultValue = undefined;
    this.defaultValueDescription = undefined;
    this.presetArg = undefined;
    this.envVar = undefined;
    this.parseArg = undefined;
    this.hidden = false;
    this.argChoices = undefined;
    this.conflictsWith = [];
    this.implied = undefined;
    this.helpGroupHeading = undefined;
  }
  default(value, description) {
    this.defaultValue = value;
    this.defaultValueDescription = description;
    return this;
  }
  preset(arg) {
    this.presetArg = arg;
    return this;
  }
  conflicts(names) {
    this.conflictsWith = this.conflictsWith.concat(names);
    return this;
  }
  implies(impliedOptionValues) {
    let newImplied = impliedOptionValues;
    if (typeof impliedOptionValues === "string") {
      newImplied = { [impliedOptionValues]: true };
    }
    this.implied = Object.assign(this.implied || {}, newImplied);
    return this;
  }
  env(name) {
    this.envVar = name;
    return this;
  }
  argParser(fn) {
    this.parseArg = fn;
    return this;
  }
  makeOptionMandatory(mandatory = true) {
    this.mandatory = !!mandatory;
    return this;
  }
  hideHelp(hide = true) {
    this.hidden = !!hide;
    return this;
  }
  _collectValue(value, previous) {
    if (previous === this.defaultValue || !Array.isArray(previous)) {
      return [value];
    }
    previous.push(value);
    return previous;
  }
  choices(values) {
    this.argChoices = values.slice();
    this.parseArg = (arg, previous) => {
      if (!this.argChoices.includes(arg)) {
        throw new InvalidArgumentError(`Allowed choices are ${this.argChoices.join(", ")}.`);
      }
      if (this.variadic) {
        return this._collectValue(arg, previous);
      }
      return arg;
    };
    return this;
  }
  name() {
    if (this.long) {
      return this.long.replace(/^--/, "");
    }
    return this.short.replace(/^-/, "");
  }
  attributeName() {
    if (this.negate) {
      return camelcase(this.name().replace(/^no-/, ""));
    }
    return camelcase(this.name());
  }
  helpGroup(heading) {
    this.helpGroupHeading = heading;
    return this;
  }
  is(arg) {
    return this.short === arg || this.long === arg;
  }
  isBoolean() {
    return !this.required && !this.optional && !this.negate;
  }
}

class DualOptions {
  constructor(options) {
    this.positiveOptions = new Map;
    this.negativeOptions = new Map;
    this.dualOptions = new Set;
    options.forEach((option) => {
      if (option.negate) {
        this.negativeOptions.set(option.attributeName(), option);
      } else {
        this.positiveOptions.set(option.attributeName(), option);
      }
    });
    this.negativeOptions.forEach((value, key) => {
      if (this.positiveOptions.has(key)) {
        this.dualOptions.add(key);
      }
    });
  }
  valueFromOption(value, option) {
    const optionKey = option.attributeName();
    if (!this.dualOptions.has(optionKey))
      return true;
    const preset = this.negativeOptions.get(optionKey).presetArg;
    const negativeValue = preset !== undefined ? preset : false;
    return option.negate === (negativeValue === value);
  }
}
function camelcase(str) {
  return str.split("-").reduce((str2, word) => {
    return str2 + word[0].toUpperCase() + word.slice(1);
  });
}
function splitOptionFlags(flags) {
  let shortFlag;
  let longFlag;
  const shortFlagExp = /^-[^-]$/;
  const longFlagExp = /^--[^-]/;
  const flagParts = flags.split(/[ |,]+/).concat("guard");
  if (shortFlagExp.test(flagParts[0]))
    shortFlag = flagParts.shift();
  if (longFlagExp.test(flagParts[0]))
    longFlag = flagParts.shift();
  if (!shortFlag && shortFlagExp.test(flagParts[0]))
    shortFlag = flagParts.shift();
  if (!shortFlag && longFlagExp.test(flagParts[0])) {
    shortFlag = longFlag;
    longFlag = flagParts.shift();
  }
  if (flagParts[0].startsWith("-")) {
    const unsupportedFlag = flagParts[0];
    const baseError = `option creation failed due to '${unsupportedFlag}' in option flags '${flags}'`;
    if (/^-[^-][^-]/.test(unsupportedFlag))
      throw new Error(`${baseError}
- a short flag is a single dash and a single character
  - either use a single dash and a single character (for a short flag)
  - or use a double dash for a long option (and can have two, like '--ws, --workspace')`);
    if (shortFlagExp.test(unsupportedFlag))
      throw new Error(`${baseError}
- too many short flags`);
    if (longFlagExp.test(unsupportedFlag))
      throw new Error(`${baseError}
- too many long flags`);
    throw new Error(`${baseError}
- unrecognised flag format`);
  }
  if (shortFlag === undefined && longFlag === undefined)
    throw new Error(`option creation failed due to no flags found in '${flags}'.`);
  return { shortFlag, longFlag };
}

// ../../node_modules/.bun/commander@15.0.0/node_modules/commander/lib/suggestSimilar.js
var maxDistance = 3;
function editDistance(a, b) {
  if (Math.abs(a.length - b.length) > maxDistance)
    return Math.max(a.length, b.length);
  const d = [];
  for (let i = 0;i <= a.length; i++) {
    d[i] = [i];
  }
  for (let j = 0;j <= b.length; j++) {
    d[0][j] = j;
  }
  for (let j = 1;j <= b.length; j++) {
    for (let i = 1;i <= a.length; i++) {
      let cost;
      if (a[i - 1] === b[j - 1]) {
        cost = 0;
      } else {
        cost = 1;
      }
      d[i][j] = Math.min(d[i - 1][j] + 1, d[i][j - 1] + 1, d[i - 1][j - 1] + cost);
      if (i > 1 && j > 1 && a[i - 1] === b[j - 2] && a[i - 2] === b[j - 1]) {
        d[i][j] = Math.min(d[i][j], d[i - 2][j - 2] + 1);
      }
    }
  }
  return d[a.length][b.length];
}
function suggestSimilar(word, candidates) {
  if (!candidates || candidates.length === 0)
    return "";
  candidates = Array.from(new Set(candidates));
  const searchingOptions = word.startsWith("--");
  if (searchingOptions) {
    word = word.slice(2);
    candidates = candidates.map((candidate) => candidate.slice(2));
  }
  let similar = [];
  let bestDistance = maxDistance;
  const minSimilarity = 0.4;
  candidates.forEach((candidate) => {
    if (candidate.length <= 1)
      return;
    const distance = editDistance(word, candidate);
    const length = Math.max(word.length, candidate.length);
    const similarity = (length - distance) / length;
    if (similarity > minSimilarity) {
      if (distance < bestDistance) {
        bestDistance = distance;
        similar = [candidate];
      } else if (distance === bestDistance) {
        similar.push(candidate);
      }
    }
  });
  similar.sort((a, b) => a.localeCompare(b));
  if (searchingOptions) {
    similar = similar.map((candidate) => `--${candidate}`);
  }
  if (similar.length > 1) {
    return `
(Did you mean one of ${similar.join(", ")}?)`;
  }
  if (similar.length === 1) {
    return `
(Did you mean ${similar[0]}?)`;
  }
  return "";
}

// ../../node_modules/.bun/commander@15.0.0/node_modules/commander/lib/command.js
class Command extends EventEmitter {
  constructor(name) {
    super();
    this.commands = [];
    this.options = [];
    this.parent = null;
    this._allowUnknownOption = false;
    this._allowExcessArguments = false;
    this.registeredArguments = [];
    this._args = this.registeredArguments;
    this.args = [];
    this.rawArgs = [];
    this.processedArgs = [];
    this._scriptPath = null;
    this._name = name || "";
    this._optionValues = {};
    this._optionValueSources = {};
    this._storeOptionsAsProperties = false;
    this._actionHandler = null;
    this._executableHandler = false;
    this._executableFile = null;
    this._executableDir = null;
    this._defaultCommandName = null;
    this._exitCallback = null;
    this._aliases = [];
    this._combineFlagAndOptionalValue = true;
    this._description = "";
    this._summary = "";
    this._argsDescription = undefined;
    this._enablePositionalOptions = false;
    this._passThroughOptions = false;
    this._lifeCycleHooks = {};
    this._showHelpAfterError = false;
    this._showSuggestionAfterError = true;
    this._savedState = null;
    this._outputConfiguration = {
      writeOut: (str) => process2.stdout.write(str),
      writeErr: (str) => process2.stderr.write(str),
      outputError: (str, write) => write(str),
      getOutHelpWidth: () => process2.stdout.isTTY ? process2.stdout.columns : undefined,
      getErrHelpWidth: () => process2.stderr.isTTY ? process2.stderr.columns : undefined,
      getOutHasColors: () => useColor() ?? (process2.stdout.isTTY && process2.stdout.hasColors?.()),
      getErrHasColors: () => useColor() ?? (process2.stderr.isTTY && process2.stderr.hasColors?.()),
      stripColor: (str) => stripVTControlCharacters2(str)
    };
    this._hidden = false;
    this._helpOption = undefined;
    this._addImplicitHelpCommand = undefined;
    this._helpCommand = undefined;
    this._helpConfiguration = {};
    this._helpGroupHeading = undefined;
    this._defaultCommandGroup = undefined;
    this._defaultOptionGroup = undefined;
  }
  copyInheritedSettings(sourceCommand) {
    this._outputConfiguration = sourceCommand._outputConfiguration;
    this._helpOption = sourceCommand._helpOption;
    this._helpCommand = sourceCommand._helpCommand;
    this._helpConfiguration = sourceCommand._helpConfiguration;
    this._exitCallback = sourceCommand._exitCallback;
    this._storeOptionsAsProperties = sourceCommand._storeOptionsAsProperties;
    this._combineFlagAndOptionalValue = sourceCommand._combineFlagAndOptionalValue;
    this._allowExcessArguments = sourceCommand._allowExcessArguments;
    this._enablePositionalOptions = sourceCommand._enablePositionalOptions;
    this._showHelpAfterError = sourceCommand._showHelpAfterError;
    this._showSuggestionAfterError = sourceCommand._showSuggestionAfterError;
    return this;
  }
  _getCommandAndAncestors() {
    const result = [];
    for (let command = this;command; command = command.parent) {
      result.push(command);
    }
    return result;
  }
  command(nameAndArgs, actionOptsOrExecDesc, execOpts) {
    let desc = actionOptsOrExecDesc;
    let opts = execOpts;
    if (typeof desc === "object" && desc !== null) {
      opts = desc;
      desc = null;
    }
    opts = opts || {};
    const [, name, args] = nameAndArgs.match(/([^ ]+) *(.*)/);
    const cmd = this.createCommand(name);
    if (desc) {
      cmd.description(desc);
      cmd._executableHandler = true;
    }
    if (opts.isDefault)
      this._defaultCommandName = cmd._name;
    cmd._hidden = !!(opts.noHelp || opts.hidden);
    cmd._executableFile = opts.executableFile || null;
    if (args)
      cmd.arguments(args);
    this._registerCommand(cmd);
    cmd.parent = this;
    cmd.copyInheritedSettings(this);
    if (desc)
      return this;
    return cmd;
  }
  createCommand(name) {
    return new Command(name);
  }
  createHelp() {
    return Object.assign(new Help, this.configureHelp());
  }
  configureHelp(configuration) {
    if (configuration === undefined)
      return this._helpConfiguration;
    this._helpConfiguration = configuration;
    return this;
  }
  configureOutput(configuration) {
    if (configuration === undefined)
      return this._outputConfiguration;
    this._outputConfiguration = {
      ...this._outputConfiguration,
      ...configuration
    };
    return this;
  }
  showHelpAfterError(displayHelp = true) {
    if (typeof displayHelp !== "string")
      displayHelp = !!displayHelp;
    this._showHelpAfterError = displayHelp;
    return this;
  }
  showSuggestionAfterError(displaySuggestion = true) {
    this._showSuggestionAfterError = !!displaySuggestion;
    return this;
  }
  addCommand(cmd, opts) {
    if (!cmd._name) {
      throw new Error(`Command passed to .addCommand() must have a name
- specify the name in Command constructor or using .name()`);
    }
    opts = opts || {};
    if (opts.isDefault)
      this._defaultCommandName = cmd._name;
    if (opts.noHelp || opts.hidden)
      cmd._hidden = true;
    this._registerCommand(cmd);
    cmd.parent = this;
    cmd._checkForBrokenPassThrough();
    return this;
  }
  createArgument(name, description) {
    return new Argument(name, description);
  }
  argument(name, description, parseArg, defaultValue) {
    const argument = this.createArgument(name, description);
    if (typeof parseArg === "function") {
      argument.default(defaultValue).argParser(parseArg);
    } else {
      argument.default(parseArg);
    }
    this.addArgument(argument);
    return this;
  }
  arguments(names) {
    names.trim().split(/ +/).forEach((detail) => {
      this.argument(detail);
    });
    return this;
  }
  addArgument(argument) {
    const previousArgument = this.registeredArguments.slice(-1)[0];
    if (previousArgument?.variadic) {
      throw new Error(`only the last argument can be variadic '${previousArgument.name()}'`);
    }
    if (argument.required && argument.defaultValue !== undefined && argument.parseArg === undefined) {
      throw new Error(`a default value for a required argument is never used: '${argument.name()}'`);
    }
    this.registeredArguments.push(argument);
    return this;
  }
  helpCommand(enableOrNameAndArgs, description) {
    if (typeof enableOrNameAndArgs === "boolean") {
      this._addImplicitHelpCommand = enableOrNameAndArgs;
      if (enableOrNameAndArgs && this._defaultCommandGroup) {
        this._initCommandGroup(this._getHelpCommand());
      }
      return this;
    }
    const nameAndArgs = enableOrNameAndArgs ?? "help [command]";
    const [, helpName, helpArgs] = nameAndArgs.match(/([^ ]+) *(.*)/);
    const helpDescription = description ?? "display help for command";
    const helpCommand = this.createCommand(helpName);
    helpCommand.helpOption(false);
    if (helpArgs)
      helpCommand.arguments(helpArgs);
    if (helpDescription)
      helpCommand.description(helpDescription);
    this._addImplicitHelpCommand = true;
    this._helpCommand = helpCommand;
    if (enableOrNameAndArgs || description)
      this._initCommandGroup(helpCommand);
    return this;
  }
  addHelpCommand(helpCommand, deprecatedDescription) {
    if (typeof helpCommand !== "object") {
      this.helpCommand(helpCommand, deprecatedDescription);
      return this;
    }
    this._addImplicitHelpCommand = true;
    this._helpCommand = helpCommand;
    this._initCommandGroup(helpCommand);
    return this;
  }
  _getHelpCommand() {
    const hasImplicitHelpCommand = this._addImplicitHelpCommand ?? (this.commands.length && !this._actionHandler && !this._findCommand("help"));
    if (hasImplicitHelpCommand) {
      if (this._helpCommand === undefined) {
        this.helpCommand(undefined, undefined);
      }
      return this._helpCommand;
    }
    return null;
  }
  hook(event, listener) {
    const allowedValues = ["preSubcommand", "preAction", "postAction"];
    if (!allowedValues.includes(event)) {
      throw new Error(`Unexpected value for event passed to hook : '${event}'.
Expecting one of '${allowedValues.join("', '")}'`);
    }
    if (this._lifeCycleHooks[event]) {
      this._lifeCycleHooks[event].push(listener);
    } else {
      this._lifeCycleHooks[event] = [listener];
    }
    return this;
  }
  exitOverride(fn) {
    if (fn) {
      this._exitCallback = fn;
    } else {
      this._exitCallback = (err) => {
        if (err.code !== "commander.executeSubCommandAsync") {
          throw err;
        }
      };
    }
    return this;
  }
  _exit(exitCode, code, message) {
    if (this._exitCallback) {
      this._exitCallback(new CommanderError(exitCode, code, message));
    }
    process2.exit(exitCode);
  }
  action(fn) {
    const listener = (args) => {
      const expectedArgsCount = this.registeredArguments.length;
      const actionArgs = args.slice(0, expectedArgsCount);
      if (this._storeOptionsAsProperties) {
        actionArgs[expectedArgsCount] = this;
      } else {
        actionArgs[expectedArgsCount] = this.opts();
      }
      actionArgs.push(this);
      return fn.apply(this, actionArgs);
    };
    this._actionHandler = listener;
    return this;
  }
  createOption(flags, description) {
    return new Option(flags, description);
  }
  _callParseArg(target, value, previous, invalidArgumentMessage) {
    try {
      return target.parseArg(value, previous);
    } catch (err) {
      if (err.code === "commander.invalidArgument") {
        const message = `${invalidArgumentMessage} ${err.message}`;
        this.error(message, { exitCode: err.exitCode, code: err.code });
      }
      throw err;
    }
  }
  _registerOption(option) {
    const matchingOption = option.short && this._findOption(option.short) || option.long && this._findOption(option.long);
    if (matchingOption) {
      const matchingFlag = option.long && this._findOption(option.long) ? option.long : option.short;
      throw new Error(`Cannot add option '${option.flags}'${this._name && ` to command '${this._name}'`} due to conflicting flag '${matchingFlag}'
-  already used by option '${matchingOption.flags}'`);
    }
    this._initOptionGroup(option);
    this.options.push(option);
  }
  _registerCommand(command) {
    const knownBy = (cmd) => {
      return [cmd.name()].concat(cmd.aliases());
    };
    const alreadyUsed = knownBy(command).find((name) => this._findCommand(name));
    if (alreadyUsed) {
      const existingCmd = knownBy(this._findCommand(alreadyUsed)).join("|");
      const newCmd = knownBy(command).join("|");
      throw new Error(`cannot add command '${newCmd}' as already have command '${existingCmd}'`);
    }
    this._initCommandGroup(command);
    this.commands.push(command);
  }
  addOption(option) {
    this._registerOption(option);
    const oname = option.name();
    const name = option.attributeName();
    if (option.defaultValue !== undefined) {
      this.setOptionValueWithSource(name, option.defaultValue, "default");
    }
    const handleOptionValue = (val, invalidValueMessage, valueSource) => {
      if (val == null && option.presetArg !== undefined) {
        val = option.presetArg;
      }
      const oldValue = this.getOptionValue(name);
      if (val !== null && option.parseArg) {
        val = this._callParseArg(option, val, oldValue, invalidValueMessage);
      } else if (val !== null && option.variadic) {
        val = option._collectValue(val, oldValue);
      }
      if (val == null) {
        if (option.negate) {
          val = false;
        } else if (option.isBoolean() || option.optional) {
          val = true;
        } else {
          val = "";
        }
      }
      this.setOptionValueWithSource(name, val, valueSource);
    };
    this.on("option:" + oname, (val) => {
      const invalidValueMessage = `error: option '${option.flags}' argument '${val}' is invalid.`;
      handleOptionValue(val, invalidValueMessage, "cli");
    });
    if (option.envVar) {
      this.on("optionEnv:" + oname, (val) => {
        const invalidValueMessage = `error: option '${option.flags}' value '${val}' from env '${option.envVar}' is invalid.`;
        handleOptionValue(val, invalidValueMessage, "env");
      });
    }
    return this;
  }
  _optionEx(config, flags, description, fn, defaultValue) {
    if (typeof flags === "object" && flags instanceof Option) {
      throw new Error("To add an Option object use addOption() instead of option() or requiredOption()");
    }
    const option = this.createOption(flags, description);
    option.makeOptionMandatory(!!config.mandatory);
    if (typeof fn === "function") {
      option.default(defaultValue).argParser(fn);
    } else if (fn instanceof RegExp) {
      const regex = fn;
      fn = (val, def) => {
        const m = regex.exec(val);
        return m ? m[0] : def;
      };
      option.default(defaultValue).argParser(fn);
    } else {
      option.default(fn);
    }
    return this.addOption(option);
  }
  option(flags, description, parseArg, defaultValue) {
    return this._optionEx({}, flags, description, parseArg, defaultValue);
  }
  requiredOption(flags, description, parseArg, defaultValue) {
    return this._optionEx({ mandatory: true }, flags, description, parseArg, defaultValue);
  }
  combineFlagAndOptionalValue(combine = true) {
    this._combineFlagAndOptionalValue = !!combine;
    return this;
  }
  allowUnknownOption(allowUnknown = true) {
    this._allowUnknownOption = !!allowUnknown;
    return this;
  }
  allowExcessArguments(allowExcess = true) {
    this._allowExcessArguments = !!allowExcess;
    return this;
  }
  enablePositionalOptions(positional = true) {
    this._enablePositionalOptions = !!positional;
    return this;
  }
  passThroughOptions(passThrough = true) {
    this._passThroughOptions = !!passThrough;
    this._checkForBrokenPassThrough();
    return this;
  }
  _checkForBrokenPassThrough() {
    if (this.parent && this._passThroughOptions && !this.parent._enablePositionalOptions) {
      throw new Error(`passThroughOptions cannot be used for '${this._name}' without turning on enablePositionalOptions for parent command(s)`);
    }
  }
  storeOptionsAsProperties(storeAsProperties = true) {
    if (this.options.length) {
      throw new Error("call .storeOptionsAsProperties() before adding options");
    }
    if (Object.keys(this._optionValues).length) {
      throw new Error("call .storeOptionsAsProperties() before setting option values");
    }
    this._storeOptionsAsProperties = !!storeAsProperties;
    return this;
  }
  getOptionValue(key) {
    if (this._storeOptionsAsProperties) {
      return this[key];
    }
    return this._optionValues[key];
  }
  setOptionValue(key, value) {
    return this.setOptionValueWithSource(key, value, undefined);
  }
  setOptionValueWithSource(key, value, source) {
    if (this._storeOptionsAsProperties) {
      this[key] = value;
    } else {
      this._optionValues[key] = value;
    }
    this._optionValueSources[key] = source;
    return this;
  }
  getOptionValueSource(key) {
    return this._optionValueSources[key];
  }
  getOptionValueSourceWithGlobals(key) {
    let source;
    this._getCommandAndAncestors().forEach((cmd) => {
      if (cmd.getOptionValueSource(key) !== undefined) {
        source = cmd.getOptionValueSource(key);
      }
    });
    return source;
  }
  _prepareUserArgs(argv, parseOptions) {
    if (argv !== undefined && !Array.isArray(argv)) {
      throw new Error("first parameter to parse must be array or undefined");
    }
    parseOptions = parseOptions || {};
    if (argv === undefined && parseOptions.from === undefined) {
      if (process2.versions?.electron) {
        parseOptions.from = "electron";
      }
      const execArgv = process2.execArgv ?? [];
      if (execArgv.includes("-e") || execArgv.includes("--eval") || execArgv.includes("-p") || execArgv.includes("--print")) {
        parseOptions.from = "eval";
      }
    }
    if (argv === undefined) {
      argv = process2.argv;
    }
    this.rawArgs = argv.slice();
    let userArgs;
    switch (parseOptions.from) {
      case undefined:
      case "node":
        this._scriptPath = argv[1];
        userArgs = argv.slice(2);
        break;
      case "electron":
        if (process2.defaultApp) {
          this._scriptPath = argv[1];
          userArgs = argv.slice(2);
        } else {
          userArgs = argv.slice(1);
        }
        break;
      case "user":
        userArgs = argv.slice(0);
        break;
      case "eval":
        userArgs = argv.slice(1);
        break;
      default:
        throw new Error(`unexpected parse option { from: '${parseOptions.from}' }`);
    }
    if (!this._name && this._scriptPath)
      this.nameFromFilename(this._scriptPath);
    this._name = this._name || "program";
    return userArgs;
  }
  parse(argv, parseOptions) {
    this._prepareForParse();
    const userArgs = this._prepareUserArgs(argv, parseOptions);
    this._parseCommand([], userArgs);
    return this;
  }
  async parseAsync(argv, parseOptions) {
    this._prepareForParse();
    const userArgs = this._prepareUserArgs(argv, parseOptions);
    await this._parseCommand([], userArgs);
    return this;
  }
  _prepareForParse() {
    if (this._savedState === null) {
      this.options.filter((option) => option.negate && option.defaultValue === undefined && this.getOptionValue(option.attributeName()) === undefined).forEach((option) => {
        const positiveLongFlag = option.long.replace(/^--no-/, "--");
        if (!this._findOption(positiveLongFlag)) {
          this.setOptionValueWithSource(option.attributeName(), true, "default");
        }
      });
      this.saveStateBeforeParse();
    } else {
      this.restoreStateBeforeParse();
    }
  }
  saveStateBeforeParse() {
    this._savedState = {
      _name: this._name,
      _optionValues: { ...this._optionValues },
      _optionValueSources: { ...this._optionValueSources }
    };
  }
  restoreStateBeforeParse() {
    if (this._storeOptionsAsProperties)
      throw new Error(`Can not call parse again when storeOptionsAsProperties is true.
- either make a new Command for each call to parse, or stop storing options as properties`);
    this._name = this._savedState._name;
    this._scriptPath = null;
    this.rawArgs = [];
    this._optionValues = { ...this._savedState._optionValues };
    this._optionValueSources = { ...this._savedState._optionValueSources };
    this.args = [];
    this.processedArgs = [];
  }
  _checkForMissingExecutable(executableFile, executableDir, subcommandName) {
    if (fs.existsSync(executableFile))
      return;
    const executableDirMessage = executableDir ? `searched for local subcommand relative to directory '${executableDir}'` : "no directory for search for local subcommand, use .executableDir() to supply a custom directory";
    const executableMissing = `'${executableFile}' does not exist
 - if '${subcommandName}' is not meant to be an executable command, remove description parameter from '.command()' and use '.description()' instead
 - if the default executable name is not suitable, use the executableFile option to supply a custom name or path
 - ${executableDirMessage}`;
    throw new Error(executableMissing);
  }
  _executeSubCommand(subcommand, args) {
    args = args.slice();
    const sourceExt = [".js", ".ts", ".tsx", ".mjs", ".cjs"];
    function findFile(baseDir, baseName) {
      const localBin = path.resolve(baseDir, baseName);
      if (fs.existsSync(localBin))
        return localBin;
      if (sourceExt.includes(path.extname(baseName)))
        return;
      const foundExt = sourceExt.find((ext) => fs.existsSync(`${localBin}${ext}`));
      if (foundExt)
        return `${localBin}${foundExt}`;
      return;
    }
    this._checkForMissingMandatoryOptions();
    this._checkForConflictingOptions();
    let executableFile = subcommand._executableFile || `${this._name}-${subcommand._name}`;
    let executableDir = this._executableDir || "";
    if (this._scriptPath) {
      let resolvedScriptPath;
      try {
        resolvedScriptPath = fs.realpathSync(this._scriptPath);
      } catch {
        resolvedScriptPath = this._scriptPath;
      }
      executableDir = path.resolve(path.dirname(resolvedScriptPath), executableDir);
    }
    if (executableDir) {
      let localFile = findFile(executableDir, executableFile);
      if (!localFile && !subcommand._executableFile && this._scriptPath) {
        const legacyName = path.basename(this._scriptPath, path.extname(this._scriptPath));
        if (legacyName !== this._name) {
          localFile = findFile(executableDir, `${legacyName}-${subcommand._name}`);
        }
      }
      executableFile = localFile || executableFile;
    }
    const launchWithNode = sourceExt.includes(path.extname(executableFile));
    let proc;
    if (process2.platform !== "win32") {
      if (launchWithNode) {
        args.unshift(executableFile);
        args = incrementNodeInspectorPort(process2.execArgv).concat(args);
        proc = childProcess.spawn(process2.argv[0], args, { stdio: "inherit" });
      } else {
        proc = childProcess.spawn(executableFile, args, { stdio: "inherit" });
      }
    } else {
      this._checkForMissingExecutable(executableFile, executableDir, subcommand._name);
      args.unshift(executableFile);
      args = incrementNodeInspectorPort(process2.execArgv).concat(args);
      proc = childProcess.spawn(process2.execPath, args, { stdio: "inherit" });
    }
    if (!proc.killed) {
      const signals = ["SIGUSR1", "SIGUSR2", "SIGTERM", "SIGINT", "SIGHUP"];
      signals.forEach((signal) => {
        process2.on(signal, () => {
          if (proc.killed === false && proc.exitCode === null) {
            proc.kill(signal);
          }
        });
      });
    }
    const exitCallback = this._exitCallback;
    proc.on("close", (code) => {
      code = code ?? 1;
      if (!exitCallback) {
        process2.exit(code);
      } else {
        exitCallback(new CommanderError(code, "commander.executeSubCommandAsync", "(close)"));
      }
    });
    proc.on("error", (err) => {
      if (err.code === "ENOENT") {
        this._checkForMissingExecutable(executableFile, executableDir, subcommand._name);
      } else if (err.code === "EACCES") {
        throw new Error(`'${executableFile}' not executable`);
      }
      if (!exitCallback) {
        process2.exit(1);
      } else {
        const wrappedError = new CommanderError(1, "commander.executeSubCommandAsync", "(error)");
        wrappedError.nestedError = err;
        exitCallback(wrappedError);
      }
    });
    this.runningCommand = proc;
  }
  _dispatchSubcommand(commandName, operands, unknown) {
    const subCommand = this._findCommand(commandName);
    if (!subCommand)
      this.help({ error: true });
    subCommand._prepareForParse();
    let promiseChain;
    promiseChain = this._chainOrCallSubCommandHook(promiseChain, subCommand, "preSubcommand");
    promiseChain = this._chainOrCall(promiseChain, () => {
      if (subCommand._executableHandler) {
        this._executeSubCommand(subCommand, operands.concat(unknown));
      } else {
        return subCommand._parseCommand(operands, unknown);
      }
    });
    return promiseChain;
  }
  _dispatchHelpCommand(subcommandName) {
    if (!subcommandName) {
      this.help();
    }
    const subCommand = this._findCommand(subcommandName);
    if (subCommand && !subCommand._executableHandler) {
      subCommand.help();
    }
    return this._dispatchSubcommand(subcommandName, [], [this._getHelpOption()?.long ?? this._getHelpOption()?.short ?? "--help"]);
  }
  _checkNumberOfArguments() {
    this.registeredArguments.forEach((arg, i) => {
      if (arg.required && this.args[i] == null) {
        this.missingArgument(arg.name());
      }
    });
    if (this.registeredArguments.length > 0 && this.registeredArguments[this.registeredArguments.length - 1].variadic) {
      return;
    }
    if (this.args.length > this.registeredArguments.length) {
      this._excessArguments(this.args);
    }
  }
  _processArguments() {
    const myParseArg = (argument, value, previous) => {
      let parsedValue = value;
      if (value !== null && argument.parseArg) {
        const invalidValueMessage = `error: command-argument value '${value}' is invalid for argument '${argument.name()}'.`;
        parsedValue = this._callParseArg(argument, value, previous, invalidValueMessage);
      }
      return parsedValue;
    };
    this._checkNumberOfArguments();
    const processedArgs = [];
    this.registeredArguments.forEach((declaredArg, index) => {
      let value = declaredArg.defaultValue;
      if (declaredArg.variadic) {
        if (index < this.args.length) {
          value = this.args.slice(index);
          if (declaredArg.parseArg) {
            value = value.reduce((processed, v) => {
              return myParseArg(declaredArg, v, processed);
            }, declaredArg.defaultValue);
          }
        } else if (value === undefined) {
          value = [];
        }
      } else if (index < this.args.length) {
        value = this.args[index];
        if (declaredArg.parseArg) {
          value = myParseArg(declaredArg, value, declaredArg.defaultValue);
        }
      }
      processedArgs[index] = value;
    });
    this.processedArgs = processedArgs;
  }
  _chainOrCall(promise, fn) {
    if (promise?.then && typeof promise.then === "function") {
      return promise.then(() => fn());
    }
    return fn();
  }
  _chainOrCallHooks(promise, event) {
    let result = promise;
    const hooks = [];
    this._getCommandAndAncestors().reverse().filter((cmd) => cmd._lifeCycleHooks[event] !== undefined).forEach((hookedCommand) => {
      hookedCommand._lifeCycleHooks[event].forEach((callback) => {
        hooks.push({ hookedCommand, callback });
      });
    });
    if (event === "postAction") {
      hooks.reverse();
    }
    hooks.forEach((hookDetail) => {
      result = this._chainOrCall(result, () => {
        return hookDetail.callback(hookDetail.hookedCommand, this);
      });
    });
    return result;
  }
  _chainOrCallSubCommandHook(promise, subCommand, event) {
    let result = promise;
    if (this._lifeCycleHooks[event] !== undefined) {
      this._lifeCycleHooks[event].forEach((hook) => {
        result = this._chainOrCall(result, () => {
          return hook(this, subCommand);
        });
      });
    }
    return result;
  }
  _parseCommand(operands, unknown) {
    const parsed = this.parseOptions(unknown);
    this._parseOptionsEnv();
    this._parseOptionsImplied();
    operands = operands.concat(parsed.operands);
    unknown = parsed.unknown;
    this.args = operands.concat(unknown);
    if (operands && this._findCommand(operands[0])) {
      return this._dispatchSubcommand(operands[0], operands.slice(1), unknown);
    }
    if (this._getHelpCommand() && operands[0] === this._getHelpCommand().name()) {
      return this._dispatchHelpCommand(operands[1]);
    }
    if (this._defaultCommandName) {
      this._outputHelpIfRequested(unknown);
      return this._dispatchSubcommand(this._defaultCommandName, operands, unknown);
    }
    if (this.commands.length && this.args.length === 0 && !this._actionHandler && !this._defaultCommandName) {
      this.help({ error: true });
    }
    this._outputHelpIfRequested(parsed.unknown);
    this._checkForMissingMandatoryOptions();
    this._checkForConflictingOptions();
    const checkForUnknownOptions = () => {
      if (parsed.unknown.length > 0) {
        this.unknownOption(parsed.unknown[0]);
      }
    };
    const commandEvent = `command:${this.name()}`;
    if (this._actionHandler) {
      checkForUnknownOptions();
      this._processArguments();
      let promiseChain;
      promiseChain = this._chainOrCallHooks(promiseChain, "preAction");
      promiseChain = this._chainOrCall(promiseChain, () => this._actionHandler(this.processedArgs));
      if (this.parent) {
        promiseChain = this._chainOrCall(promiseChain, () => {
          this.parent.emit(commandEvent, operands, unknown);
        });
      }
      promiseChain = this._chainOrCallHooks(promiseChain, "postAction");
      return promiseChain;
    }
    if (this.parent?.listenerCount(commandEvent)) {
      checkForUnknownOptions();
      this._processArguments();
      this.parent.emit(commandEvent, operands, unknown);
    } else if (operands.length) {
      if (this._findCommand("*")) {
        return this._dispatchSubcommand("*", operands, unknown);
      }
      if (this.listenerCount("command:*")) {
        this.emit("command:*", operands, unknown);
      } else if (this.commands.length) {
        this.unknownCommand();
      } else {
        checkForUnknownOptions();
        this._processArguments();
      }
    } else if (this.commands.length) {
      checkForUnknownOptions();
      this.help({ error: true });
    } else {
      checkForUnknownOptions();
      this._processArguments();
    }
  }
  _findCommand(name) {
    if (!name)
      return;
    return this.commands.find((cmd) => cmd._name === name || cmd._aliases.includes(name));
  }
  _findOption(arg) {
    return this.options.find((option) => option.is(arg));
  }
  _checkForMissingMandatoryOptions() {
    this._getCommandAndAncestors().forEach((cmd) => {
      cmd.options.forEach((anOption) => {
        if (anOption.mandatory && cmd.getOptionValue(anOption.attributeName()) === undefined) {
          cmd.missingMandatoryOptionValue(anOption);
        }
      });
    });
  }
  _checkForConflictingLocalOptions() {
    const definedNonDefaultOptions = this.options.filter((option) => {
      const optionKey = option.attributeName();
      if (this.getOptionValue(optionKey) === undefined) {
        return false;
      }
      return this.getOptionValueSource(optionKey) !== "default";
    });
    const optionsWithConflicting = definedNonDefaultOptions.filter((option) => option.conflictsWith.length > 0);
    optionsWithConflicting.forEach((option) => {
      const conflictingAndDefined = definedNonDefaultOptions.find((defined) => option.conflictsWith.includes(defined.attributeName()));
      if (conflictingAndDefined) {
        this._conflictingOption(option, conflictingAndDefined);
      }
    });
  }
  _checkForConflictingOptions() {
    this._getCommandAndAncestors().forEach((cmd) => {
      cmd._checkForConflictingLocalOptions();
    });
  }
  parseOptions(args) {
    const operands = [];
    const unknown = [];
    let dest = operands;
    function maybeOption(arg) {
      return arg.length > 1 && arg[0] === "-";
    }
    const negativeNumberArg = (arg) => {
      if (!/^-(\d+|\d*\.\d+)(e[+-]?\d+)?$/.test(arg))
        return false;
      return !this._getCommandAndAncestors().some((cmd) => cmd.options.map((opt) => opt.short).some((short) => /^-\d$/.test(short)));
    };
    let activeVariadicOption = null;
    let activeGroup = null;
    let i = 0;
    while (i < args.length || activeGroup) {
      const arg = activeGroup ?? args[i++];
      activeGroup = null;
      if (arg === "--") {
        if (dest === unknown)
          dest.push(arg);
        dest.push(...args.slice(i));
        break;
      }
      if (activeVariadicOption && (!maybeOption(arg) || negativeNumberArg(arg))) {
        this.emit(`option:${activeVariadicOption.name()}`, arg);
        continue;
      }
      activeVariadicOption = null;
      if (maybeOption(arg)) {
        const option = this._findOption(arg);
        if (option) {
          if (option.required) {
            const value = args[i++];
            if (value === undefined)
              this.optionMissingArgument(option);
            this.emit(`option:${option.name()}`, value);
          } else if (option.optional) {
            let value = null;
            if (i < args.length && (!maybeOption(args[i]) || negativeNumberArg(args[i]))) {
              value = args[i++];
            }
            this.emit(`option:${option.name()}`, value);
          } else {
            this.emit(`option:${option.name()}`);
          }
          activeVariadicOption = option.variadic ? option : null;
          continue;
        }
      }
      if (arg.length > 2 && arg[0] === "-" && arg[1] !== "-") {
        const option = this._findOption(`-${arg[1]}`);
        if (option) {
          if (option.required || option.optional && this._combineFlagAndOptionalValue) {
            this.emit(`option:${option.name()}`, arg.slice(2));
          } else {
            this.emit(`option:${option.name()}`);
            activeGroup = `-${arg.slice(2)}`;
          }
          continue;
        }
      }
      if (/^--[^=]+=/.test(arg)) {
        const index = arg.indexOf("=");
        const option = this._findOption(arg.slice(0, index));
        if (option && (option.required || option.optional)) {
          this.emit(`option:${option.name()}`, arg.slice(index + 1));
          continue;
        }
      }
      if (dest === operands && maybeOption(arg) && !(this.commands.length === 0 && negativeNumberArg(arg))) {
        dest = unknown;
      }
      if ((this._enablePositionalOptions || this._passThroughOptions) && operands.length === 0 && unknown.length === 0) {
        if (this._findCommand(arg)) {
          operands.push(arg);
          unknown.push(...args.slice(i));
          break;
        } else if (this._getHelpCommand() && arg === this._getHelpCommand().name()) {
          operands.push(arg, ...args.slice(i));
          break;
        } else if (this._defaultCommandName) {
          unknown.push(arg, ...args.slice(i));
          break;
        }
      }
      if (this._passThroughOptions) {
        dest.push(arg, ...args.slice(i));
        break;
      }
      dest.push(arg);
    }
    return { operands, unknown };
  }
  opts() {
    if (this._storeOptionsAsProperties) {
      const result = {};
      const len = this.options.length;
      for (let i = 0;i < len; i++) {
        const key = this.options[i].attributeName();
        result[key] = key === this._versionOptionName ? this._version : this[key];
      }
      return result;
    }
    return this._optionValues;
  }
  optsWithGlobals() {
    return this._getCommandAndAncestors().reduce((combinedOptions, cmd) => Object.assign(combinedOptions, cmd.opts()), {});
  }
  error(message, errorOptions) {
    this._outputConfiguration.outputError(`${message}
`, this._outputConfiguration.writeErr);
    if (typeof this._showHelpAfterError === "string") {
      this._outputConfiguration.writeErr(`${this._showHelpAfterError}
`);
    } else if (this._showHelpAfterError) {
      this._outputConfiguration.writeErr(`
`);
      this.outputHelp({ error: true });
    }
    const config = errorOptions || {};
    const exitCode = config.exitCode || 1;
    const code = config.code || "commander.error";
    this._exit(exitCode, code, message);
  }
  _parseOptionsEnv() {
    this.options.forEach((option) => {
      if (option.envVar && option.envVar in process2.env) {
        const optionKey = option.attributeName();
        if (this.getOptionValue(optionKey) === undefined || ["default", "config", "env"].includes(this.getOptionValueSource(optionKey))) {
          if (option.required || option.optional) {
            this.emit(`optionEnv:${option.name()}`, process2.env[option.envVar]);
          } else {
            this.emit(`optionEnv:${option.name()}`);
          }
        }
      }
    });
  }
  _parseOptionsImplied() {
    const dualHelper = new DualOptions(this.options);
    const hasCustomOptionValue = (optionKey) => {
      return this.getOptionValue(optionKey) !== undefined && !["default", "implied"].includes(this.getOptionValueSource(optionKey));
    };
    this.options.filter((option) => option.implied !== undefined && hasCustomOptionValue(option.attributeName()) && dualHelper.valueFromOption(this.getOptionValue(option.attributeName()), option)).forEach((option) => {
      Object.keys(option.implied).filter((impliedKey) => !hasCustomOptionValue(impliedKey)).forEach((impliedKey) => {
        this.setOptionValueWithSource(impliedKey, option.implied[impliedKey], "implied");
      });
    });
  }
  missingArgument(name) {
    const message = `error: missing required argument '${name}'`;
    this.error(message, { code: "commander.missingArgument" });
  }
  optionMissingArgument(option) {
    const message = `error: option '${option.flags}' argument missing`;
    this.error(message, { code: "commander.optionMissingArgument" });
  }
  missingMandatoryOptionValue(option) {
    const message = `error: required option '${option.flags}' not specified`;
    this.error(message, { code: "commander.missingMandatoryOptionValue" });
  }
  _conflictingOption(option, conflictingOption) {
    const findBestOptionFromValue = (option2) => {
      const optionKey = option2.attributeName();
      const optionValue = this.getOptionValue(optionKey);
      const negativeOption = this.options.find((target) => target.negate && optionKey === target.attributeName());
      const positiveOption = this.options.find((target) => !target.negate && optionKey === target.attributeName());
      if (negativeOption && (negativeOption.presetArg === undefined && optionValue === false || negativeOption.presetArg !== undefined && optionValue === negativeOption.presetArg)) {
        return negativeOption;
      }
      return positiveOption || option2;
    };
    const getErrorMessage = (option2) => {
      const bestOption = findBestOptionFromValue(option2);
      const optionKey = bestOption.attributeName();
      const source = this.getOptionValueSource(optionKey);
      if (source === "env") {
        return `environment variable '${bestOption.envVar}'`;
      }
      return `option '${bestOption.flags}'`;
    };
    const message = `error: ${getErrorMessage(option)} cannot be used with ${getErrorMessage(conflictingOption)}`;
    this.error(message, { code: "commander.conflictingOption" });
  }
  unknownOption(flag) {
    if (this._allowUnknownOption)
      return;
    let suggestion = "";
    if (flag.startsWith("--") && this._showSuggestionAfterError) {
      let candidateFlags = [];
      let command = this;
      do {
        const moreFlags = command.createHelp().visibleOptions(command).filter((option) => option.long).map((option) => option.long);
        candidateFlags = candidateFlags.concat(moreFlags);
        command = command.parent;
      } while (command && !command._enablePositionalOptions);
      suggestion = suggestSimilar(flag, candidateFlags);
    }
    const message = `error: unknown option '${flag}'${suggestion}`;
    this.error(message, { code: "commander.unknownOption" });
  }
  _excessArguments(receivedArgs) {
    if (this._allowExcessArguments)
      return;
    const expected = this.registeredArguments.length;
    const s = expected === 1 ? "" : "s";
    const received = receivedArgs.length;
    const forSubcommand = this.parent ? ` for '${this.name()}'` : "";
    const details = receivedArgs.join(", ");
    const message = `error: too many arguments${forSubcommand}. Expected ${expected} argument${s} but got ${received}: ${details}.`;
    this.error(message, { code: "commander.excessArguments" });
  }
  unknownCommand() {
    const unknownName = this.args[0];
    let suggestion = "";
    if (this._showSuggestionAfterError) {
      const candidateNames = [];
      this.createHelp().visibleCommands(this).forEach((command) => {
        candidateNames.push(command.name());
        if (command.alias())
          candidateNames.push(command.alias());
      });
      suggestion = suggestSimilar(unknownName, candidateNames);
    }
    const message = `error: unknown command '${unknownName}'${suggestion}`;
    this.error(message, { code: "commander.unknownCommand" });
  }
  version(str, flags, description) {
    if (str === undefined)
      return this._version;
    this._version = str;
    flags = flags || "-V, --version";
    description = description || "output the version number";
    const versionOption = this.createOption(flags, description);
    this._versionOptionName = versionOption.attributeName();
    this._registerOption(versionOption);
    this.on("option:" + versionOption.name(), () => {
      this._outputConfiguration.writeOut(`${str}
`);
      this._exit(0, "commander.version", str);
    });
    return this;
  }
  description(str, argsDescription) {
    if (str === undefined && argsDescription === undefined)
      return this._description;
    this._description = str;
    if (argsDescription) {
      this._argsDescription = argsDescription;
    }
    return this;
  }
  summary(str) {
    if (str === undefined)
      return this._summary;
    this._summary = str;
    return this;
  }
  alias(alias) {
    if (alias === undefined)
      return this._aliases[0];
    let command = this;
    if (this.commands.length !== 0 && this.commands[this.commands.length - 1]._executableHandler) {
      command = this.commands[this.commands.length - 1];
    }
    if (alias === command._name)
      throw new Error("Command alias can't be the same as its name");
    const matchingCommand = this.parent?._findCommand(alias);
    if (matchingCommand) {
      const existingCmd = [matchingCommand.name()].concat(matchingCommand.aliases()).join("|");
      throw new Error(`cannot add alias '${alias}' to command '${this.name()}' as already have command '${existingCmd}'`);
    }
    command._aliases.push(alias);
    return this;
  }
  aliases(aliases) {
    if (aliases === undefined)
      return this._aliases;
    aliases.forEach((alias) => this.alias(alias));
    return this;
  }
  usage(str) {
    if (str === undefined) {
      if (this._usage)
        return this._usage;
      const args = this.registeredArguments.map((arg) => {
        return humanReadableArgName(arg);
      });
      return [].concat(this.options.length || this._helpOption !== null ? "[options]" : [], this.commands.length ? "[command]" : [], this.registeredArguments.length ? args : []).join(" ");
    }
    this._usage = str;
    return this;
  }
  name(str) {
    if (str === undefined)
      return this._name;
    this._name = str;
    return this;
  }
  helpGroup(heading) {
    if (heading === undefined)
      return this._helpGroupHeading ?? "";
    this._helpGroupHeading = heading;
    return this;
  }
  commandsGroup(heading) {
    if (heading === undefined)
      return this._defaultCommandGroup ?? "";
    this._defaultCommandGroup = heading;
    return this;
  }
  optionsGroup(heading) {
    if (heading === undefined)
      return this._defaultOptionGroup ?? "";
    this._defaultOptionGroup = heading;
    return this;
  }
  _initOptionGroup(option) {
    if (this._defaultOptionGroup && !option.helpGroupHeading)
      option.helpGroup(this._defaultOptionGroup);
  }
  _initCommandGroup(cmd) {
    if (this._defaultCommandGroup && !cmd.helpGroup())
      cmd.helpGroup(this._defaultCommandGroup);
  }
  nameFromFilename(filename) {
    this._name = path.basename(filename, path.extname(filename));
    return this;
  }
  executableDir(path2) {
    if (path2 === undefined)
      return this._executableDir;
    this._executableDir = path2;
    return this;
  }
  helpInformation(contextOptions) {
    const helper = this.createHelp();
    const context = this._getOutputContext(contextOptions);
    helper.prepareContext({
      error: context.error,
      helpWidth: context.helpWidth,
      outputHasColors: context.hasColors
    });
    const text = helper.formatHelp(this, helper);
    if (context.hasColors)
      return text;
    return this._outputConfiguration.stripColor(text);
  }
  _getOutputContext(contextOptions) {
    contextOptions = contextOptions || {};
    const error = !!contextOptions.error;
    let baseWrite;
    let hasColors;
    let helpWidth;
    if (error) {
      baseWrite = (str) => this._outputConfiguration.writeErr(str);
      hasColors = this._outputConfiguration.getErrHasColors();
      helpWidth = this._outputConfiguration.getErrHelpWidth();
    } else {
      baseWrite = (str) => this._outputConfiguration.writeOut(str);
      hasColors = this._outputConfiguration.getOutHasColors();
      helpWidth = this._outputConfiguration.getOutHelpWidth();
    }
    const write = (str) => {
      if (!hasColors)
        str = this._outputConfiguration.stripColor(str);
      return baseWrite(str);
    };
    return { error, write, hasColors, helpWidth };
  }
  outputHelp(contextOptions) {
    let deprecatedCallback;
    if (typeof contextOptions === "function") {
      deprecatedCallback = contextOptions;
      contextOptions = undefined;
    }
    const outputContext = this._getOutputContext(contextOptions);
    const eventContext = {
      error: outputContext.error,
      write: outputContext.write,
      command: this
    };
    this._getCommandAndAncestors().reverse().forEach((command) => command.emit("beforeAllHelp", eventContext));
    this.emit("beforeHelp", eventContext);
    let helpInformation = this.helpInformation({ error: outputContext.error });
    if (deprecatedCallback) {
      helpInformation = deprecatedCallback(helpInformation);
      if (typeof helpInformation !== "string" && !Buffer.isBuffer(helpInformation)) {
        throw new Error("outputHelp callback must return a string or a Buffer");
      }
    }
    outputContext.write(helpInformation);
    if (this._getHelpOption()?.long) {
      this.emit(this._getHelpOption().long);
    }
    this.emit("afterHelp", eventContext);
    this._getCommandAndAncestors().forEach((command) => command.emit("afterAllHelp", eventContext));
  }
  helpOption(flags, description) {
    if (typeof flags === "boolean") {
      if (flags) {
        if (this._helpOption === null)
          this._helpOption = undefined;
        if (this._defaultOptionGroup) {
          this._initOptionGroup(this._getHelpOption());
        }
      } else {
        this._helpOption = null;
      }
      return this;
    }
    this._helpOption = this.createOption(flags ?? "-h, --help", description ?? "display help for command");
    if (flags || description)
      this._initOptionGroup(this._helpOption);
    return this;
  }
  _getHelpOption() {
    if (this._helpOption === undefined) {
      this.helpOption(undefined, undefined);
    }
    return this._helpOption;
  }
  addHelpOption(option) {
    this._helpOption = option;
    this._initOptionGroup(option);
    return this;
  }
  help(contextOptions) {
    this.outputHelp(contextOptions);
    let exitCode = Number(process2.exitCode ?? 0);
    if (exitCode === 0 && contextOptions && typeof contextOptions !== "function" && contextOptions.error) {
      exitCode = 1;
    }
    this._exit(exitCode, "commander.help", "(outputHelp)");
  }
  addHelpText(position, text) {
    const allowedValues = ["beforeAll", "before", "after", "afterAll"];
    if (!allowedValues.includes(position)) {
      throw new Error(`Unexpected value for position to addHelpText.
Expecting one of '${allowedValues.join("', '")}'`);
    }
    const helpEvent = `${position}Help`;
    this.on(helpEvent, (context) => {
      let helpStr;
      if (typeof text === "function") {
        helpStr = text({ error: context.error, command: context.command });
      } else {
        helpStr = text;
      }
      if (helpStr) {
        context.write(`${helpStr}
`);
      }
    });
    return this;
  }
  _outputHelpIfRequested(args) {
    const helpOption = this._getHelpOption();
    const helpRequested = helpOption && args.find((arg) => helpOption.is(arg));
    if (helpRequested) {
      this.outputHelp();
      this._exit(0, "commander.helpDisplayed", "(outputHelp)");
    }
  }
}
function incrementNodeInspectorPort(args) {
  return args.map((arg) => {
    if (!arg.startsWith("--inspect")) {
      return arg;
    }
    let debugOption;
    let debugHost = "127.0.0.1";
    let debugPort = "9229";
    let match;
    if ((match = arg.match(/^(--inspect(-brk)?)$/)) !== null) {
      debugOption = match[1];
    } else if ((match = arg.match(/^(--inspect(-brk|-port)?)=([^:]+)$/)) !== null) {
      debugOption = match[1];
      if (/^\d+$/.test(match[3])) {
        debugPort = match[3];
      } else {
        debugHost = match[3];
      }
    } else if ((match = arg.match(/^(--inspect(-brk|-port)?)=([^:]+):(\d+)$/)) !== null) {
      debugOption = match[1];
      debugHost = match[3];
      debugPort = match[4];
    }
    if (debugOption && debugPort !== "0") {
      return `${debugOption}=${debugHost}:${parseInt(debugPort) + 1}`;
    }
    return arg;
  });
}
function useColor() {
  if (process2.env.NO_COLOR || process2.env.FORCE_COLOR === "0" || process2.env.FORCE_COLOR === "false")
    return false;
  if (process2.env.FORCE_COLOR || process2.env.CLICOLOR_FORCE !== undefined)
    return true;
  return;
}

// ../../node_modules/.bun/commander@15.0.0/node_modules/commander/index.js
var program = new Command;

// ../../node_modules/.bun/chalk@5.6.2/node_modules/chalk/source/vendor/ansi-styles/index.js
var ANSI_BACKGROUND_OFFSET = 10;
var wrapAnsi16 = (offset = 0) => (code) => `\x1B[${code + offset}m`;
var wrapAnsi256 = (offset = 0) => (code) => `\x1B[${38 + offset};5;${code}m`;
var wrapAnsi16m = (offset = 0) => (red, green, blue) => `\x1B[${38 + offset};2;${red};${green};${blue}m`;
var styles = {
  modifier: {
    reset: [0, 0],
    bold: [1, 22],
    dim: [2, 22],
    italic: [3, 23],
    underline: [4, 24],
    overline: [53, 55],
    inverse: [7, 27],
    hidden: [8, 28],
    strikethrough: [9, 29]
  },
  color: {
    black: [30, 39],
    red: [31, 39],
    green: [32, 39],
    yellow: [33, 39],
    blue: [34, 39],
    magenta: [35, 39],
    cyan: [36, 39],
    white: [37, 39],
    blackBright: [90, 39],
    gray: [90, 39],
    grey: [90, 39],
    redBright: [91, 39],
    greenBright: [92, 39],
    yellowBright: [93, 39],
    blueBright: [94, 39],
    magentaBright: [95, 39],
    cyanBright: [96, 39],
    whiteBright: [97, 39]
  },
  bgColor: {
    bgBlack: [40, 49],
    bgRed: [41, 49],
    bgGreen: [42, 49],
    bgYellow: [43, 49],
    bgBlue: [44, 49],
    bgMagenta: [45, 49],
    bgCyan: [46, 49],
    bgWhite: [47, 49],
    bgBlackBright: [100, 49],
    bgGray: [100, 49],
    bgGrey: [100, 49],
    bgRedBright: [101, 49],
    bgGreenBright: [102, 49],
    bgYellowBright: [103, 49],
    bgBlueBright: [104, 49],
    bgMagentaBright: [105, 49],
    bgCyanBright: [106, 49],
    bgWhiteBright: [107, 49]
  }
};
var modifierNames = Object.keys(styles.modifier);
var foregroundColorNames = Object.keys(styles.color);
var backgroundColorNames = Object.keys(styles.bgColor);
var colorNames = [...foregroundColorNames, ...backgroundColorNames];
function assembleStyles() {
  const codes = new Map;
  for (const [groupName, group] of Object.entries(styles)) {
    for (const [styleName, style] of Object.entries(group)) {
      styles[styleName] = {
        open: `\x1B[${style[0]}m`,
        close: `\x1B[${style[1]}m`
      };
      group[styleName] = styles[styleName];
      codes.set(style[0], style[1]);
    }
    Object.defineProperty(styles, groupName, {
      value: group,
      enumerable: false
    });
  }
  Object.defineProperty(styles, "codes", {
    value: codes,
    enumerable: false
  });
  styles.color.close = "\x1B[39m";
  styles.bgColor.close = "\x1B[49m";
  styles.color.ansi = wrapAnsi16();
  styles.color.ansi256 = wrapAnsi256();
  styles.color.ansi16m = wrapAnsi16m();
  styles.bgColor.ansi = wrapAnsi16(ANSI_BACKGROUND_OFFSET);
  styles.bgColor.ansi256 = wrapAnsi256(ANSI_BACKGROUND_OFFSET);
  styles.bgColor.ansi16m = wrapAnsi16m(ANSI_BACKGROUND_OFFSET);
  Object.defineProperties(styles, {
    rgbToAnsi256: {
      value(red, green, blue) {
        if (red === green && green === blue) {
          if (red < 8) {
            return 16;
          }
          if (red > 248) {
            return 231;
          }
          return Math.round((red - 8) / 247 * 24) + 232;
        }
        return 16 + 36 * Math.round(red / 255 * 5) + 6 * Math.round(green / 255 * 5) + Math.round(blue / 255 * 5);
      },
      enumerable: false
    },
    hexToRgb: {
      value(hex) {
        const matches = /[a-f\d]{6}|[a-f\d]{3}/i.exec(hex.toString(16));
        if (!matches) {
          return [0, 0, 0];
        }
        let [colorString] = matches;
        if (colorString.length === 3) {
          colorString = [...colorString].map((character) => character + character).join("");
        }
        const integer = Number.parseInt(colorString, 16);
        return [
          integer >> 16 & 255,
          integer >> 8 & 255,
          integer & 255
        ];
      },
      enumerable: false
    },
    hexToAnsi256: {
      value: (hex) => styles.rgbToAnsi256(...styles.hexToRgb(hex)),
      enumerable: false
    },
    ansi256ToAnsi: {
      value(code) {
        if (code < 8) {
          return 30 + code;
        }
        if (code < 16) {
          return 90 + (code - 8);
        }
        let red;
        let green;
        let blue;
        if (code >= 232) {
          red = ((code - 232) * 10 + 8) / 255;
          green = red;
          blue = red;
        } else {
          code -= 16;
          const remainder = code % 36;
          red = Math.floor(code / 36) / 5;
          green = Math.floor(remainder / 6) / 5;
          blue = remainder % 6 / 5;
        }
        const value = Math.max(red, green, blue) * 2;
        if (value === 0) {
          return 30;
        }
        let result = 30 + (Math.round(blue) << 2 | Math.round(green) << 1 | Math.round(red));
        if (value === 2) {
          result += 60;
        }
        return result;
      },
      enumerable: false
    },
    rgbToAnsi: {
      value: (red, green, blue) => styles.ansi256ToAnsi(styles.rgbToAnsi256(red, green, blue)),
      enumerable: false
    },
    hexToAnsi: {
      value: (hex) => styles.ansi256ToAnsi(styles.hexToAnsi256(hex)),
      enumerable: false
    }
  });
  return styles;
}
var ansiStyles = assembleStyles();
var ansi_styles_default = ansiStyles;

// ../../node_modules/.bun/chalk@5.6.2/node_modules/chalk/source/vendor/supports-color/index.js
import process3 from "node:process";
import os from "node:os";
import tty from "node:tty";
function hasFlag(flag, argv = globalThis.Deno ? globalThis.Deno.args : process3.argv) {
  const prefix = flag.startsWith("-") ? "" : flag.length === 1 ? "-" : "--";
  const position = argv.indexOf(prefix + flag);
  const terminatorPosition = argv.indexOf("--");
  return position !== -1 && (terminatorPosition === -1 || position < terminatorPosition);
}
var { env } = process3;
var flagForceColor;
if (hasFlag("no-color") || hasFlag("no-colors") || hasFlag("color=false") || hasFlag("color=never")) {
  flagForceColor = 0;
} else if (hasFlag("color") || hasFlag("colors") || hasFlag("color=true") || hasFlag("color=always")) {
  flagForceColor = 1;
}
function envForceColor() {
  if ("FORCE_COLOR" in env) {
    if (env.FORCE_COLOR === "true") {
      return 1;
    }
    if (env.FORCE_COLOR === "false") {
      return 0;
    }
    return env.FORCE_COLOR.length === 0 ? 1 : Math.min(Number.parseInt(env.FORCE_COLOR, 10), 3);
  }
}
function translateLevel(level) {
  if (level === 0) {
    return false;
  }
  return {
    level,
    hasBasic: true,
    has256: level >= 2,
    has16m: level >= 3
  };
}
function _supportsColor(haveStream, { streamIsTTY, sniffFlags = true } = {}) {
  const noFlagForceColor = envForceColor();
  if (noFlagForceColor !== undefined) {
    flagForceColor = noFlagForceColor;
  }
  const forceColor = sniffFlags ? flagForceColor : noFlagForceColor;
  if (forceColor === 0) {
    return 0;
  }
  if (sniffFlags) {
    if (hasFlag("color=16m") || hasFlag("color=full") || hasFlag("color=truecolor")) {
      return 3;
    }
    if (hasFlag("color=256")) {
      return 2;
    }
  }
  if ("TF_BUILD" in env && "AGENT_NAME" in env) {
    return 1;
  }
  if (haveStream && !streamIsTTY && forceColor === undefined) {
    return 0;
  }
  const min = forceColor || 0;
  if (env.TERM === "dumb") {
    return min;
  }
  if (process3.platform === "win32") {
    const osRelease = os.release().split(".");
    if (Number(osRelease[0]) >= 10 && Number(osRelease[2]) >= 10586) {
      return Number(osRelease[2]) >= 14931 ? 3 : 2;
    }
    return 1;
  }
  if ("CI" in env) {
    if (["GITHUB_ACTIONS", "GITEA_ACTIONS", "CIRCLECI"].some((key) => (key in env))) {
      return 3;
    }
    if (["TRAVIS", "APPVEYOR", "GITLAB_CI", "BUILDKITE", "DRONE"].some((sign) => (sign in env)) || env.CI_NAME === "codeship") {
      return 1;
    }
    return min;
  }
  if ("TEAMCITY_VERSION" in env) {
    return /^(9\.(0*[1-9]\d*)\.|\d{2,}\.)/.test(env.TEAMCITY_VERSION) ? 1 : 0;
  }
  if (env.COLORTERM === "truecolor") {
    return 3;
  }
  if (env.TERM === "xterm-kitty") {
    return 3;
  }
  if (env.TERM === "xterm-ghostty") {
    return 3;
  }
  if (env.TERM === "wezterm") {
    return 3;
  }
  if ("TERM_PROGRAM" in env) {
    const version = Number.parseInt((env.TERM_PROGRAM_VERSION || "").split(".")[0], 10);
    switch (env.TERM_PROGRAM) {
      case "iTerm.app": {
        return version >= 3 ? 3 : 2;
      }
      case "Apple_Terminal": {
        return 2;
      }
    }
  }
  if (/-256(color)?$/i.test(env.TERM)) {
    return 2;
  }
  if (/^screen|^xterm|^vt100|^vt220|^rxvt|color|ansi|cygwin|linux/i.test(env.TERM)) {
    return 1;
  }
  if ("COLORTERM" in env) {
    return 1;
  }
  return min;
}
function createSupportsColor(stream, options = {}) {
  const level = _supportsColor(stream, {
    streamIsTTY: stream && stream.isTTY,
    ...options
  });
  return translateLevel(level);
}
var supportsColor = {
  stdout: createSupportsColor({ isTTY: tty.isatty(1) }),
  stderr: createSupportsColor({ isTTY: tty.isatty(2) })
};
var supports_color_default = supportsColor;

// ../../node_modules/.bun/chalk@5.6.2/node_modules/chalk/source/utilities.js
function stringReplaceAll(string, substring, replacer) {
  let index = string.indexOf(substring);
  if (index === -1) {
    return string;
  }
  const substringLength = substring.length;
  let endIndex = 0;
  let returnValue = "";
  do {
    returnValue += string.slice(endIndex, index) + substring + replacer;
    endIndex = index + substringLength;
    index = string.indexOf(substring, endIndex);
  } while (index !== -1);
  returnValue += string.slice(endIndex);
  return returnValue;
}
function stringEncaseCRLFWithFirstIndex(string, prefix, postfix, index) {
  let endIndex = 0;
  let returnValue = "";
  do {
    const gotCR = string[index - 1] === "\r";
    returnValue += string.slice(endIndex, gotCR ? index - 1 : index) + prefix + (gotCR ? `\r
` : `
`) + postfix;
    endIndex = index + 1;
    index = string.indexOf(`
`, endIndex);
  } while (index !== -1);
  returnValue += string.slice(endIndex);
  return returnValue;
}

// ../../node_modules/.bun/chalk@5.6.2/node_modules/chalk/source/index.js
var { stdout: stdoutColor, stderr: stderrColor } = supports_color_default;
var GENERATOR = Symbol("GENERATOR");
var STYLER = Symbol("STYLER");
var IS_EMPTY = Symbol("IS_EMPTY");
var levelMapping = [
  "ansi",
  "ansi",
  "ansi256",
  "ansi16m"
];
var styles2 = Object.create(null);
var applyOptions = (object, options = {}) => {
  if (options.level && !(Number.isInteger(options.level) && options.level >= 0 && options.level <= 3)) {
    throw new Error("The `level` option should be an integer from 0 to 3");
  }
  const colorLevel = stdoutColor ? stdoutColor.level : 0;
  object.level = options.level === undefined ? colorLevel : options.level;
};
var chalkFactory = (options) => {
  const chalk = (...strings) => strings.join(" ");
  applyOptions(chalk, options);
  Object.setPrototypeOf(chalk, createChalk.prototype);
  return chalk;
};
function createChalk(options) {
  return chalkFactory(options);
}
Object.setPrototypeOf(createChalk.prototype, Function.prototype);
for (const [styleName, style] of Object.entries(ansi_styles_default)) {
  styles2[styleName] = {
    get() {
      const builder = createBuilder(this, createStyler(style.open, style.close, this[STYLER]), this[IS_EMPTY]);
      Object.defineProperty(this, styleName, { value: builder });
      return builder;
    }
  };
}
styles2.visible = {
  get() {
    const builder = createBuilder(this, this[STYLER], true);
    Object.defineProperty(this, "visible", { value: builder });
    return builder;
  }
};
var getModelAnsi = (model, level, type, ...arguments_) => {
  if (model === "rgb") {
    if (level === "ansi16m") {
      return ansi_styles_default[type].ansi16m(...arguments_);
    }
    if (level === "ansi256") {
      return ansi_styles_default[type].ansi256(ansi_styles_default.rgbToAnsi256(...arguments_));
    }
    return ansi_styles_default[type].ansi(ansi_styles_default.rgbToAnsi(...arguments_));
  }
  if (model === "hex") {
    return getModelAnsi("rgb", level, type, ...ansi_styles_default.hexToRgb(...arguments_));
  }
  return ansi_styles_default[type][model](...arguments_);
};
var usedModels = ["rgb", "hex", "ansi256"];
for (const model of usedModels) {
  styles2[model] = {
    get() {
      const { level } = this;
      return function(...arguments_) {
        const styler = createStyler(getModelAnsi(model, levelMapping[level], "color", ...arguments_), ansi_styles_default.color.close, this[STYLER]);
        return createBuilder(this, styler, this[IS_EMPTY]);
      };
    }
  };
  const bgModel = "bg" + model[0].toUpperCase() + model.slice(1);
  styles2[bgModel] = {
    get() {
      const { level } = this;
      return function(...arguments_) {
        const styler = createStyler(getModelAnsi(model, levelMapping[level], "bgColor", ...arguments_), ansi_styles_default.bgColor.close, this[STYLER]);
        return createBuilder(this, styler, this[IS_EMPTY]);
      };
    }
  };
}
var proto = Object.defineProperties(() => {}, {
  ...styles2,
  level: {
    enumerable: true,
    get() {
      return this[GENERATOR].level;
    },
    set(level) {
      this[GENERATOR].level = level;
    }
  }
});
var createStyler = (open, close, parent) => {
  let openAll;
  let closeAll;
  if (parent === undefined) {
    openAll = open;
    closeAll = close;
  } else {
    openAll = parent.openAll + open;
    closeAll = close + parent.closeAll;
  }
  return {
    open,
    close,
    openAll,
    closeAll,
    parent
  };
};
var createBuilder = (self, _styler, _isEmpty) => {
  const builder = (...arguments_) => applyStyle(builder, arguments_.length === 1 ? "" + arguments_[0] : arguments_.join(" "));
  Object.setPrototypeOf(builder, proto);
  builder[GENERATOR] = self;
  builder[STYLER] = _styler;
  builder[IS_EMPTY] = _isEmpty;
  return builder;
};
var applyStyle = (self, string) => {
  if (self.level <= 0 || !string) {
    return self[IS_EMPTY] ? "" : string;
  }
  let styler = self[STYLER];
  if (styler === undefined) {
    return string;
  }
  const { openAll, closeAll } = styler;
  if (string.includes("\x1B")) {
    while (styler !== undefined) {
      string = stringReplaceAll(string, styler.close, styler.open);
      styler = styler.parent;
    }
  }
  const lfIndex = string.indexOf(`
`);
  if (lfIndex !== -1) {
    string = stringEncaseCRLFWithFirstIndex(string, closeAll, openAll, lfIndex);
  }
  return openAll + string + closeAll;
};
Object.defineProperties(createChalk.prototype, styles2);
var chalk = createChalk();
var chalkStderr = createChalk({ level: stderrColor ? stderrColor.level : 0 });
var source_default = chalk;

// src/cli.ts
import { existsSync as existsSync3, readdirSync as readdirSync3, readFileSync as readFileSync3, mkdirSync, writeFileSync as writeFileSync2 } from "fs";
import { join as join3, resolve as resolve2 } from "path";

// src/index.ts
import { readFileSync, readdirSync } from "fs";
import { join } from "path";
function tokenize(text) {
  return text.toLowerCase().replace(/[^a-z0-9\s]/g, " ").split(/\s+/).filter((t) => t.length > 2 && !stopWords.has(t));
}
var stopWords = new Set([
  "the",
  "and",
  "for",
  "are",
  "but",
  "not",
  "you",
  "all",
  "can",
  "had",
  "her",
  "was",
  "one",
  "our",
  "out",
  "has",
  "have",
  "been",
  "some",
  "than",
  "that",
  "this",
  "with",
  "from",
  "they",
  "will",
  "would",
  "which",
  "their",
  "there",
  "about",
  "into",
  "each",
  "said",
  "does",
  "also",
  "after",
  "before",
  "other",
  "more",
  "only",
  "over",
  "such",
  "when",
  "where",
  "what",
  "who",
  "how",
  "then",
  "just",
  "very",
  "much",
  "well",
  "should",
  "could",
  "through",
  "between"
]);
function buildIndex(projectDir, projectName) {
  const files = readdirSync(projectDir).filter((f) => f.endsWith(".dog"));
  const entries = [];
  const docFreq = new Map;
  for (const file of files) {
    const content = readFileSync(join(projectDir, file), "utf-8");
    const sections = content.split(/\n(?=##\s)/);
    for (const section of sections) {
      const lines = section.split(`
`);
      const heading = lines[0]?.replace(/^#+\s*/, "") || "(root)";
      const body = lines.slice(1).join(`
`);
      if (body.trim().length < 20)
        continue;
      const terms = tokenize(body);
      if (terms.length < 5)
        continue;
      const tf = new Map;
      for (const t of terms) {
        tf.set(t, (tf.get(t) || 0) + 1);
        docFreq.set(t, (docFreq.get(t) || 0) + 1);
      }
      entries.push({
        section: heading,
        heading,
        file,
        content: body.slice(0, 500),
        vector: []
      });
    }
  }
  const vocabulary = [...docFreq.entries()].filter(([_, df2]) => df2 >= 2).sort((a, b) => b[1] - a[1]).slice(0, 1000).map(([term]) => term);
  const df = vocabulary.map((t) => docFreq.get(t) || 0);
  const N = entries.length;
  for (const entry of entries) {
    const terms = tokenize(entry.content);
    const tf = new Map;
    for (const t of terms)
      tf.set(t, (tf.get(t) || 0) + 1);
    entry.vector = vocabulary.map((term) => {
      const termFreq = tf.get(term) || 0;
      if (termFreq === 0)
        return 0;
      const docFreqTerm = docFreq.get(term) || 1;
      return termFreq / terms.length * Math.log(N / docFreqTerm);
    });
  }
  return {
    version: "1.0",
    project: projectName,
    built: new Date().toISOString(),
    entries,
    vocabulary,
    df
  };
}
function searchIndex(index, query, limit = 10) {
  const queryTerms = tokenize(query);
  const queryVec = index.vocabulary.map((term) => queryTerms.includes(term) ? 1 : 0);
  const results = index.entries.map((entry) => {
    let dot = 0;
    let entryMag = 0;
    let queryMag = 0;
    for (let j = 0;j < queryVec.length; j++) {
      dot += queryVec[j] * entry.vector[j];
      queryMag += queryVec[j] * queryVec[j];
      entryMag += entry.vector[j] * entry.vector[j];
    }
    const score = queryMag > 0 && entryMag > 0 ? dot / (Math.sqrt(queryMag) * Math.sqrt(entryMag)) : 0;
    return { entry, score };
  });
  return results.filter((r) => r.score > 0).sort((a, b) => b.score - a.score).slice(0, limit);
}

// src/cli.ts
import { homedir as homedir2 } from "os";

// src/parser.ts
function parse(source) {
  const lines = source.split(`
`);
  const errors2 = [];
  const sections = parseSections(lines, errors2);
  return { kind: "document", sections, errors: errors2 };
}
function parseToJSON(source) {
  const ast = parse(source);
  return JSON.stringify(ast, null, 2);
}
function parseSections(lines, errors2) {
  const sections = [];
  let i = 0;
  let firstHeading = lines.length;
  for (let j = 0;j < lines.length; j++) {
    if (/^##\s/.test(lines[j])) {
      firstHeading = j;
      break;
    }
  }
  const rootBlocks = parseBlocks(lines, 0, firstHeading, errors2);
  if (rootBlocks.length > 0) {
    sections.push({
      kind: "section",
      level: 1,
      heading: "(root)",
      blocks: rootBlocks,
      lineStart: 1,
      lineEnd: lines.length
    });
  }
  while (i < lines.length) {
    const line = lines[i];
    const h2 = line.match(/^##\s+(.+)/);
    const h3 = line.match(/^###\s+(.+)/);
    if (h2 || h3) {
      const level = h2 ? 2 : 3;
      const heading = (h2 || h3)[1];
      const sectionStart = i;
      i++;
      const end = findSectionEnd(lines, i, level);
      const blocks = parseBlocks(lines, sectionStart, end, errors2);
      sections.push({
        kind: "section",
        level,
        heading,
        blocks,
        lineStart: sectionStart + 1,
        lineEnd: end + 1
      });
      i = end;
    } else {
      i++;
    }
  }
  return sections;
}
function findSectionEnd(lines, start, currentLevel) {
  for (let i = start;i < lines.length; i++) {
    const line = lines[i];
    if (/^##\s/.test(line) || /^###\s/.test(line))
      return i;
  }
  return lines.length;
}
function parseBlocks(lines, start, end, errors2) {
  const blocks = [];
  let i = start;
  while (i < end) {
    const line = lines[i];
    const entityMatch = line.match(/^#{3,5}\s+Entity:\s*(.+)/);
    if (entityMatch) {
      const result = parseStructuredBlock(lines, i, end, "entity", entityMatch[1]);
      if (result) {
        blocks.push(result.node);
        i = result.nextLine;
        continue;
      }
    }
    const relMatch = line.match(/^#{3,5}\s+Relationship:\s*(.+)/);
    if (relMatch) {
      const result = parseStructuredBlock(lines, i, end, "relationship", relMatch[1]);
      if (result) {
        blocks.push(result.node);
        i = result.nextLine;
        continue;
      }
    }
    const eventMatch = line.match(/^#{3,5}\s+Event:\s*(.+)/);
    if (eventMatch) {
      const result = parseStructuredBlock(lines, i, end, "event", eventMatch[1]);
      if (result) {
        blocks.push(result.node);
        i = result.nextLine;
        continue;
      }
    }
    const predMatch = line.match(/^#{3,5}\s+Prediction:\s*(.+)/);
    if (predMatch) {
      const result = parseStructuredBlock(lines, i, end, "prediction", predMatch[1]);
      if (result) {
        blocks.push(result.node);
        i = result.nextLine;
        continue;
      }
    }
    if (lines[i].startsWith("```")) {
      let yamlEnd = i + 1;
      while (yamlEnd < end && !lines[yamlEnd].startsWith("```"))
        yamlEnd++;
      if (yamlEnd < end) {
        const yamlContent = lines.slice(i + 1, yamlEnd);
        let yaml;
        try {
          yaml = parseSimpleYAML(yamlContent);
        } catch (e) {
          if (errors2)
            errors2.push({ message: `YAML parse error: ${e.message}`, line: i + 1, context: lines[i - 1]?.trim() || "(unknown)" });
          i = yamlEnd + 1;
          continue;
        }
        if (yaml.prediction) {
          blocks.push({
            kind: "prediction",
            statement: yaml.prediction || "",
            description: yaml.description || "",
            trigger: yaml.trigger || "",
            timeframe: yaml.timeframe || "",
            confidence: yaml.confidence || 0,
            measurement: yaml.measurement || "",
            status: yaml.status || "pending",
            yaml,
            lineStart: i + 1,
            lineEnd: yamlEnd
          });
          i = yamlEnd + 1;
          continue;
        }
        if (yaml.entity) {
          blocks.push({
            kind: "entity",
            name: yaml.entity || "",
            description: yaml.description || "",
            type: yaml.type || "node",
            properties: {},
            states: Array.isArray(yaml.states) ? yaml.states : [],
            lifecycle: [],
            yaml,
            lineStart: i + 1,
            lineEnd: yamlEnd
          });
          i = yamlEnd + 1;
          continue;
        }
        if (yaml.relationship || yaml.verb) {
          blocks.push({
            kind: "relationship",
            source: yaml.source || "",
            target: yaml.target || "",
            verb: yaml.verb || "connects",
            description: yaml.description || "",
            cardinality: yaml.cardinality || "N:M",
            required: false,
            cascade: "none",
            invariants: [],
            yaml,
            lineStart: i + 1,
            lineEnd: yamlEnd
          });
          i = yamlEnd + 1;
          continue;
        }
        if (Array.isArray(yaml.relationships)) {
          for (const item of yaml.relationships) {
            blocks.push({
              kind: "relationship",
              source: item.from || item.source || "",
              target: item.to || item.target || "",
              verb: item.verb || "connects",
              description: item.description || "",
              cardinality: item.cardinality || "N:M",
              required: item.required === true,
              cascade: item.cascade || "none",
              invariants: [],
              yaml: item,
              lineStart: i + 1,
              lineEnd: yamlEnd
            });
          }
          i = yamlEnd + 1;
          continue;
        }
        if (yaml.event) {
          blocks.push({
            kind: "event",
            name: yaml.event || "",
            trigger: yaml.trigger || "",
            payload: {},
            preconditions: [],
            postconditions: [],
            sideEffects: [],
            probability: null,
            yaml,
            lineStart: i + 1,
            lineEnd: yamlEnd
          });
          i = yamlEnd + 1;
          continue;
        }
      }
    }
    if (/^\|.+\|/.test(line) && i + 1 < end && /^\|[-| ]+\|/.test(lines[i + 1])) {
      const table = parseTable(lines, i, end);
      if (table) {
        blocks.push(table);
        i = table.lineEnd;
        continue;
      }
    }
    if (lines[i].startsWith("```")) {
      let yamlEnd = i + 1;
      while (yamlEnd < end && !lines[yamlEnd].startsWith("```"))
        yamlEnd++;
      if (yamlEnd < end && yamlEnd > i + 1) {
        const yamlContent = lines.slice(i + 1, yamlEnd);
        let yaml;
        try {
          yaml = parseSimpleYAML(yamlContent);
        } catch (e) {
          if (errors2)
            errors2.push({ message: `YAML parse error: ${e.message}`, line: i + 1, context: lines[i - 1]?.trim() || "(unknown)" });
          i = yamlEnd + 1;
          continue;
        }
        const key = yaml.prediction ? "prediction" : yaml.entity ? "entity" : yaml.event ? "event" : yaml.relationship || yaml.verb ? "relationship" : Array.isArray(yaml.relationships) ? "relationship_list" : null;
        if (key) {
          if (key === "prediction") {
            blocks.push({ kind: "prediction", statement: yaml.prediction || "", description: yaml.description || "", trigger: yaml.trigger || "", timeframe: yaml.timeframe || "", confidence: yaml.confidence || 0, measurement: yaml.measurement || "", status: yaml.status || "pending", yaml, lineStart: i + 1, lineEnd: yamlEnd });
          } else if (key === "entity") {
            blocks.push({ kind: "entity", name: yaml.entity || "", description: yaml.description || "", type: yaml.type || "node", properties: {}, states: Array.isArray(yaml.states) ? yaml.states : [], lifecycle: [], yaml, lineStart: i + 1, lineEnd: yamlEnd });
          } else if (key === "event") {
            blocks.push({ kind: "event", name: yaml.event || "", trigger: yaml.trigger || "", payload: {}, preconditions: [], postconditions: [], sideEffects: [], probability: null, yaml, lineStart: i + 1, lineEnd: yamlEnd });
          } else if (key === "relationship") {
            blocks.push({ kind: "relationship", source: yaml.source || "", target: yaml.target || "", verb: yaml.verb || "connects", description: yaml.description || "", cardinality: yaml.cardinality || "N:M", required: false, cascade: "none", invariants: [], yaml, lineStart: i + 1, lineEnd: yamlEnd });
          } else if (key === "relationship_list") {
            for (const item of yaml.relationships) {
              blocks.push({ kind: "relationship", source: item.from || item.source || "", target: item.to || item.target || "", verb: item.verb || "connects", description: item.description || "", cardinality: item.cardinality || "N:M", required: item.required === true, cascade: item.cascade || "none", invariants: [], yaml: item, lineStart: i + 1, lineEnd: yamlEnd });
            }
          }
          i = yamlEnd + 1;
          continue;
        }
      }
    }
    if (lines[i].startsWith("```")) {
      let fenceEnd = i + 1;
      while (fenceEnd < end && !lines[fenceEnd].startsWith("```"))
        fenceEnd++;
      const blockEnd = fenceEnd < end ? fenceEnd + 1 : end;
      blocks.push({
        kind: "prose",
        content: lines.slice(i, blockEnd).join(`
`).trim(),
        lineStart: i + 1,
        lineEnd: blockEnd
      });
      i = blockEnd;
      continue;
    }
    const proseStart = i;
    while (i < end && !isBlockStart(lines[i]) && !lines[i].startsWith("```")) {
      i++;
    }
    const proseLines = lines.slice(proseStart, i).filter((l) => l.trim() !== "" || i === proseStart + 1);
    if (proseLines.length > 0) {
      blocks.push({
        kind: "prose",
        content: proseLines.join(`
`).trim(),
        lineStart: proseStart + 1,
        lineEnd: i
      });
    }
  }
  return blocks;
}
function isBlockStart(line) {
  return /^#{3,5}\s+(Entity|Relationship|Event|Prediction):/.test(line) || /^\|.+\|/.test(line);
}
function parseStructuredBlock(lines, start, end, kind, headerRest) {
  let i = start + 1;
  let description = "";
  while (i < end && !lines[i].startsWith("```") && !isBlockStart(lines[i])) {
    if (lines[i].trim())
      description += (description ? " " : "") + lines[i].trim();
    i++;
  }
  if (i >= end || !lines[i].startsWith("```")) {
    return {
      node: {
        kind,
        name: headerRest,
        description,
        type: "node",
        properties: {},
        states: [],
        lifecycle: [],
        yaml: {},
        lineStart: start + 1,
        lineEnd: i
      },
      nextLine: i
    };
  }
  i++;
  const yamlLines = [];
  while (i < end && !lines[i].startsWith("```")) {
    yamlLines.push(lines[i]);
    i++;
  }
  i++;
  let yaml = {};
  try {
    yaml = parseSimpleYAML(yamlLines);
  } catch (e) {
    if (errors)
      errors.push({ message: `YAML parse error: ${e.message}`, line: start + 1, context: `${kind}: ${headerRest}` });
    return {
      node: { kind: "entity", name: headerRest, description, type: "node", properties: {}, states: [], lifecycle: [], yaml: {}, lineStart: start + 1, lineEnd: i },
      nextLine: i
    };
  }
  if (kind === "entity") {
    return {
      node: buildEntityNode(headerRest, description, yaml, start, i),
      nextLine: i
    };
  }
  if (kind === "relationship") {
    return {
      node: buildRelationshipNode(headerRest, description, yaml, start, i),
      nextLine: i
    };
  }
  if (kind === "event") {
    return {
      node: buildEventNode(headerRest, description, yaml, start, i),
      nextLine: i
    };
  }
  if (kind === "prediction") {
    return {
      node: buildPredictionNode(headerRest, description, yaml, start, i),
      nextLine: i
    };
  }
  return null;
}
function buildEntityNode(name, description, yaml, lineStart, lineEnd) {
  const properties = {};
  const rawProps = yaml.properties;
  if (rawProps) {
    for (const [key, val] of Object.entries(rawProps)) {
      let resolved = val;
      if (typeof val === "string" && val.startsWith("{") && val.endsWith("}")) {
        const inline = parseInlineObject(val);
        if (inline)
          resolved = inline;
      }
      if (typeof resolved === "object" && resolved !== null) {
        properties[key] = {
          type: resolved.type || "string",
          required: resolved.required !== false,
          default: resolved.default,
          constraints: resolved.constraints,
          example: resolved.example
        };
      }
    }
  }
  const states = Array.isArray(yaml.states) ? yaml.states : [];
  const lifecycleStr = yaml.lifecycle || "";
  const lifecycleParts = lifecycleStr ? lifecycleStr.split(/\s*→\s*/).map((s) => s.trim()) : [];
  const lifecycle = [];
  for (let si = 0;si < lifecycleParts.length - 1; si++) {
    lifecycle.push(`${lifecycleParts[si]} → ${lifecycleParts[si + 1]}`);
  }
  return {
    kind: "entity",
    name,
    description: description || yaml.description || "",
    type: yaml.type || "node",
    properties,
    states,
    lifecycle,
    yaml,
    lineStart: lineStart + 1,
    lineEnd
  };
}
function buildRelationshipNode(headerRest, description, yaml, lineStart, lineEnd) {
  const parts = headerRest.split(/\s*→\s*/);
  const source = parts[0]?.trim() || yaml.source || "";
  const target = parts[1]?.trim() || yaml.target || "";
  return {
    kind: "relationship",
    source,
    target,
    verb: yaml.verb || "connects",
    description: description || yaml.description || "",
    cardinality: yaml.cardinality || "N:M",
    required: yaml.required === true,
    cascade: yaml.cascade || "none",
    invariants: Array.isArray(yaml.invariants) ? yaml.invariants : [yaml.invariants || ""],
    yaml,
    lineStart: lineStart + 1,
    lineEnd
  };
}
function buildEventNode(name, description, yaml, lineStart, lineEnd) {
  return {
    kind: "event",
    name,
    trigger: yaml.trigger || "",
    payload: yaml.payload || {},
    preconditions: Array.isArray(yaml.preconditions) ? yaml.preconditions : [],
    postconditions: Array.isArray(yaml.postconditions) ? yaml.postconditions : [],
    sideEffects: Array.isArray(yaml.sideEffects) ? yaml.sideEffects : [],
    probability: typeof yaml.probability === "number" ? yaml.probability : null,
    yaml,
    lineStart: lineStart + 1,
    lineEnd
  };
}
function buildPredictionNode(name, description, yaml, lineStart, lineEnd) {
  return {
    kind: "prediction",
    statement: name,
    description,
    trigger: yaml.trigger || "",
    timeframe: yaml.timeframe || "",
    confidence: yaml.confidence || 0,
    measurement: yaml.measurement || "",
    status: yaml.status || "pending",
    yaml,
    lineStart,
    lineEnd
  };
}
function parseTable(lines, start, end) {
  const headerLine = lines[start];
  const headers = headerLine.split("|").map((h) => h.trim()).filter(Boolean);
  let i = start + 2;
  const rows = [];
  while (i < end && lines[i].startsWith("|")) {
    const row = lines[i].split("|").map((c) => c.trim()).filter(Boolean);
    rows.push(row);
    i++;
  }
  if (headers.length === 0)
    return null;
  return {
    kind: "table",
    headers,
    rows,
    lineStart: start + 1,
    lineEnd: i
  };
}
function parseSimpleYAML(lines) {
  const result = {};
  let currentKey = "";
  let currentObj = {};
  let inNested = false;
  let nestedKey = "";
  for (const rawLine of lines) {
    const line = rawLine.trimEnd();
    if (!line.trim())
      continue;
    const topMatch = line.match(/^(\w[\w_]*):\s*(.+)?$/);
    if (topMatch && !line.startsWith("  ") && !line.startsWith("\t")) {
      if (inNested && currentKey) {
        if (Array.isArray(currentObj.__list) && currentObj.__list.length > 0) {
          result[currentKey] = currentObj.__list;
          delete currentObj.__list;
        } else {
          result[currentKey] = currentObj;
        }
        inNested = false;
        currentObj = {};
      }
      currentKey = topMatch[1];
      const value = (topMatch[2] || "").trim();
      if (value === "") {
        inNested = true;
        currentObj = {};
      } else if (value === "true") {
        result[currentKey] = true;
      } else if (value === "false") {
        result[currentKey] = false;
      } else if (/^-?\d+(\.\d+)?$/.test(value)) {
        result[currentKey] = parseFloat(value);
      } else if (value.startsWith("[") && value.endsWith("]")) {
        result[currentKey] = value.slice(1, -1).split(",").map((s) => s.trim());
      } else if (value.startsWith("null")) {
        result[currentKey] = null;
      } else {
        result[currentKey] = value;
      }
      continue;
    }
    const nestedMatch = line.match(/^\s{2}(\w[\w_]*):\s*(.+)?$/);
    if (nestedMatch && inNested) {
      nestedKey = nestedMatch[1];
      const value = (nestedMatch[2] || "").trim();
      if (value === "" || value === "{") {
        currentObj[nestedKey] = {};
      } else if (value === "true") {
        currentObj[nestedKey] = true;
      } else if (value === "false") {
        currentObj[nestedKey] = false;
      } else if (/^-?\d+(\.\d+)?$/.test(value)) {
        currentObj[nestedKey] = parseFloat(value);
      } else if (value.startsWith("[") && value.endsWith("]")) {
        currentObj[nestedKey] = value.slice(1, -1).split(",").map((s) => s.trim());
      } else {
        currentObj[nestedKey] = value;
      }
      continue;
    }
    const listMatch = line.match(/^\s{2}-(?:\s+)?(.+)?$/);
    if (listMatch && inNested && currentKey) {
      const item = (listMatch[1] || "").trim();
      if (!Array.isArray(currentObj.__list)) {
        currentObj.__list = [];
      }
      if (item.startsWith("{") && item.endsWith("}")) {
        const inline = parseInlineObject(item);
        currentObj.__list.push(inline || item);
      } else if (item === "true") {
        currentObj.__list.push(true);
      } else if (item === "false") {
        currentObj.__list.push(false);
      } else if (/^-?\d+(\.\d+)?$/.test(item)) {
        currentObj.__list.push(parseFloat(item));
      } else {
        currentObj.__list.push(item);
      }
      continue;
    }
    if (nestedMatch && !inNested) {
      const key = nestedMatch[1];
      const value = (nestedMatch[2] || "").trim();
      if (value === "true")
        result[key] = true;
      else if (value === "false")
        result[key] = false;
      else if (/^-?\d+(\.\d+)?$/.test(value))
        result[key] = parseFloat(value);
      else if (value.startsWith("[") && value.endsWith("]"))
        result[key] = value.slice(1, -1).split(",").map((s) => s.trim());
      else
        result[key] = value;
      continue;
    }
    const deepMatch = line.match(/^\s{4}(\w[\w_]*):\s*(.+)?$/);
    if (deepMatch && inNested && nestedKey && typeof currentObj[nestedKey] === "object") {
      const deepNested = currentObj[nestedKey];
      const key = deepMatch[1];
      const value = (deepMatch[2] || "").trim();
      const inlineObj = parseInlineObject(value);
      deepNested[key] = inlineObj || value;
    }
  }
  if (inNested && currentKey) {
    if (Array.isArray(currentObj.__list) && currentObj.__list.length > 0) {
      result[currentKey] = currentObj.__list;
      delete currentObj.__list;
    } else {
      result[currentKey] = currentObj;
    }
  }
  return result;
}
function parseInlineObject(value) {
  const match = value.match(/^\{(.+)\}$/);
  if (!match)
    return null;
  const obj = {};
  const pairs = match[1].split(",").map((s) => s.trim());
  for (const pair of pairs) {
    const [k, ...vParts] = pair.split(":");
    if (!k || vParts.length === 0)
      continue;
    let v = vParts.join(":").trim();
    if (v === "true")
      obj[k.trim()] = true;
    else if (v === "false")
      obj[k.trim()] = false;
    else if (/^-?\d+(\.\d+)?$/.test(v))
      obj[k.trim()] = parseFloat(v);
    else
      obj[k.trim()] = v;
  }
  return obj;
}
// src/serve.ts
import { existsSync as existsSync2, readdirSync as readdirSync2, readFileSync as readFileSync2 } from "fs";
import { join as join2, resolve } from "path";
import { homedir } from "os";
import * as readline from "readline";
function resolvePath(p) {
  if (p.startsWith("~"))
    p = join2(homedir(), p.slice(1));
  const resolved = p.startsWith("/") ? p : join2(process.cwd(), p);
  if (!p.startsWith("/") && !p.startsWith("~")) {
    const rel = resolve(process.cwd(), p);
    const cwd = process.cwd();
    const isDescendant = rel.startsWith(cwd + "/");
    const isSelf = rel === cwd;
    const isAncestor = cwd.startsWith(rel + "/");
    if (!isDescendant && !isSelf && !isAncestor) {
      throw new Error(`Path traversal blocked: ${p}`);
    }
    return rel;
  }
  return resolved;
}
var isV2 = (n) => Array.isArray(n) && typeof n[0] === "number";
var N = (dag) => dag.n || dag.nodes || [];
function edgeToObj(n, tgtIdx, v2e) {
  return { t: String(tgtIdx), v: v2e[1] || "", c: v2e[2] || "", r: v2e[3] || 0 };
}
function nodeEdges(n) {
  if (isV2(n))
    return (n[6] || []).map((e) => edgeToObj(n, e[0], e));
  return n.es || [];
}
function E(dag) {
  if (dag.e)
    return dag.e;
  const edges = [];
  const seen = new Set;
  for (const node of N(dag)) {
    for (const e of nodeEdges(node)) {
      const src = nodeId(node);
      const key = `${src}→${e.t}:${e.v}`;
      if (!seen.has(key)) {
        seen.add(key);
        edges.push({ s: src, t: e.t, v: e.v, d: e.d, c: e.c, r: e.r });
      }
    }
  }
  return edges;
}
var P = (dag) => dag.p || dag.project || "";
var nodeId = (n) => isV2(n) ? String(n[0]) : n.i || n.id || "";
var nodeName = (n) => isV2(n) ? n[1] || String(n[0]) : n.i || n.id || n.name || "";
function nodeMatches(n, value) {
  const q = (value || "").toLowerCase();
  return nodeId(n).toLowerCase() === q || nodeName(n).toLowerCase() === q;
}
var nt = (n) => isV2(n) ? n[2] || "" : n.t || n.type || "";
var nd = (n) => isV2(n) ? n[3] || "" : n.d || n.description || "";
function np(n) {
  if (isV2(n)) {
    const flat = n[4] || [];
    const obj = {};
    for (let i = 0;i < flat.length; i += 2)
      obj[flat[i]] = flat[i + 1] || "";
    return obj;
  }
  return n.p || n.properties || {};
}
var ns = (n) => isV2(n) ? n[5] || [] : n.s || n.states || [];
var nl = (n) => isV2(n) ? [] : n.l || n.lifecycle || [];
var es = (e) => e.s || e.source || "";
var et = (e) => e.t || e.target || "";
function serve(dir = ".") {
  const root = resolvePath(dir);
  const dagCache = new Map;
  function loadDags() {
    const dirs = [join2(root, "projects"), join2(root, "specs"), root];
    for (const dd of dirs) {
      if (!existsSync2(dd))
        continue;
      const projects = readdirSync2(dd, { withFileTypes: true }).filter((e) => e.isDirectory()).map((e) => e.name);
      for (const p of projects) {
        const dagFile = join2(dd, p, `${p}.dag`);
        if (existsSync2(dagFile)) {
          dagCache.set(p, JSON.parse(readFileSync2(dagFile, "utf-8")));
        }
      }
    }
    return [...dagCache.keys()];
  }
  function handleRequest(req) {
    const { id, method, params } = req;
    if (method === "initialize") {
      return {
        jsonrpc: "2.0",
        id,
        result: {
          protocolVersion: "2024-11-05",
          serverInfo: { name: "spec-serve", version: "0.1.0" },
          capabilities: { tools: {} }
        }
      };
    }
    if (method === "tools/list") {
      return {
        jsonrpc: "2.0",
        id,
        result: {
          tools: [
            { name: "getEntity", description: "Get entity with properties, states, lifecycle", inputSchema: { type: "object", properties: { project: { type: "string" }, name: { type: "string" } }, required: ["name"] } },
            { name: "traverse", description: "Traverse graph: from node, follow edges, return subgraph", inputSchema: { type: "object", properties: { project: { type: "string" }, from: { type: "string" }, depth: { type: "number", default: 1 } }, required: ["from"] } },
            { name: "search", description: "Find entities by name or type", inputSchema: { type: "object", properties: { project: { type: "string" }, q: { type: "string" }, type: { type: "string" } }, required: ["q"] } },
            { name: "listProjects", description: "List all projects", inputSchema: { type: "object", properties: {} } },
            { name: "summary", description: "Get project summary: node count, edge count, token savings", inputSchema: { type: "object", properties: { project: { type: "string" } } } },
            { name: "schema", description: "Get full property schema for an entity", inputSchema: { type: "object", properties: { project: { type: "string" }, entity: { type: "string" } }, required: ["entity"] } }
          ]
        }
      };
    }
    if (method === "tools/call") {
      const { name, arguments: args } = params;
      loadDags();
      if (name === "listProjects") {
        return { jsonrpc: "2.0", id, result: { content: [{ type: "text", text: JSON.stringify([...dagCache.keys()]) }] } };
      }
      if (name === "listNodes") {
        const dag = dagCache.get(args.project || [...dagCache.keys()][0] || "");
        if (!dag)
          return { jsonrpc: "2.0", id, error: { code: 404, message: "Project not found" } };
        return { jsonrpc: "2.0", id, result: { content: [{ type: "text", text: JSON.stringify(N(dag)) }] } };
      }
      if (name === "getEntity") {
        const dag = dagCache.get(args.project || [...dagCache.keys()][0] || "");
        if (!dag)
          return { jsonrpc: "2.0", id, error: { code: 404, message: "Project not found" } };
        const node = N(dag).find((n) => nodeMatches(n, args.name || ""));
        if (!node)
          return { jsonrpc: "2.0", id, result: { content: [{ type: "text", text: "{}" }] } };
        const idForEdges = nodeId(node);
        const edges = E(dag).filter((e) => es(e).toLowerCase() === idForEdges.toLowerCase() || et(e).toLowerCase() === idForEdges.toLowerCase());
        return { jsonrpc: "2.0", id, result: { content: [{ type: "text", text: JSON.stringify(isV2(node) ? { id: nodeId(node), name: nodeName(node), type: nt(node), description: nd(node), properties: np(node), states: ns(node), edges } : { ...node, edges }) }] } };
      }
      if (name === "traverse") {
        const dag = dagCache.get(args.project || [...dagCache.keys()][0] || "");
        if (!dag)
          return { jsonrpc: "2.0", id, error: { code: 404, message: "Project not found" } };
        const depth = Math.min(Math.max(1, args.depth || 1), 20);
        const visitedNodes = new Set;
        const visitedEdges = new Set;
        const subgraph = { nodes: [], edges: [] };
        const start = N(dag).find((n) => nodeMatches(n, args.from || ""));
        const queue = [{ id: start ? nodeId(start) : args.from, depth: 0 }];
        while (queue.length > 0) {
          const curr = queue.shift();
          const node = N(dag).find((n) => nodeMatches(n, curr.id));
          const currId = node ? nodeId(node) : curr.id;
          if (visitedNodes.has(currId) || curr.depth > depth)
            continue;
          visitedNodes.add(currId);
          if (node)
            subgraph.nodes.push(isV2(node) ? { id: nodeId(node), name: nodeName(node), type: nt(node), description: nd(node), properties: np(node), states: ns(node), edges: nodeEdges(node) } : node);
          const edges = E(dag).filter((e) => es(e).toLowerCase() === currId.toLowerCase() || et(e).toLowerCase() === currId.toLowerCase());
          for (const e of edges) {
            const edgeKey = `${es(e)}→${et(e)}`;
            if (!visitedEdges.has(edgeKey)) {
              visitedEdges.add(edgeKey);
              subgraph.edges.push(e);
            }
            const next = es(e).toLowerCase() === currId.toLowerCase() ? et(e) : es(e);
            if (!visitedNodes.has(next))
              queue.push({ id: next, depth: curr.depth + 1 });
          }
        }
        return { jsonrpc: "2.0", id, result: { content: [{ type: "text", text: JSON.stringify(subgraph) }] } };
      }
      if (name === "search") {
        const dag = dagCache.get(args.project || [...dagCache.keys()][0] || "");
        if (!dag)
          return { jsonrpc: "2.0", id, error: { code: 404, message: "Project not found" } };
        const q = (args.q || "").toLowerCase();
        const type = (args.type || "").toLowerCase();
        const results = N(dag).filter((n) => (nodeName(n).toLowerCase().includes(q) || nodeId(n).toLowerCase().includes(q) || nt(n).toLowerCase().includes(q)) && (!type || nt(n).toLowerCase().includes(type)));
        return { jsonrpc: "2.0", id, result: { content: [{ type: "text", text: JSON.stringify(results) }] } };
      }
      if (name === "summary") {
        const dag = dagCache.get(args.project || [...dagCache.keys()][0] || "");
        if (!dag)
          return { jsonrpc: "2.0", id, error: { code: 404, message: "Project not found" } };
        const tk = dag.tk || dag.tokens || {};
        const s = {
          project: P(dag),
          nodes: N(dag).length,
          edges: E(dag).length,
          version: dag.v || dag.version || (Array.isArray(dag) ? `${dag[0]}` : "unknown"),
          order: dag.o || [],
          cycles: dag.cy !== undefined ? dag.cy : null,
          savings: tk.sv || tk.savings_pct || 0,
          method: tk.m || tk.method || ""
        };
        return { jsonrpc: "2.0", id, result: { content: [{ type: "text", text: JSON.stringify(s) }] } };
      }
      if (name === "schema") {
        const dag = dagCache.get(args.project || [...dagCache.keys()][0] || "");
        if (!dag)
          return { jsonrpc: "2.0", id, error: { code: 404, message: "Project not found" } };
        const node = N(dag).find((n) => nodeMatches(n, args.entity || ""));
        if (!node)
          return { jsonrpc: "2.0", id, error: { code: 404, message: "Entity not found" } };
        return { jsonrpc: "2.0", id, result: { content: [{ type: "text", text: JSON.stringify({
          entity: nodeName(node),
          properties: np(node),
          states: ns(node),
          lifecycle: nl(node)
        }) }] } };
      }
      return { jsonrpc: "2.0", id, error: { code: 404, message: `Unknown tool: ${name}` } };
    }
    return { jsonrpc: "2.0", id, error: { code: 404, message: `Unknown method: ${method}` } };
  }
  const rl = readline.createInterface({ input: process.stdin });
  loadDags();
  console.error(`[spec-serve] Loaded ${dagCache.size} projects`);
  rl.on("line", (line) => {
    try {
      const req = JSON.parse(line);
      const res = handleRequest(req);
      process.stdout.write(JSON.stringify(res) + `
`);
    } catch (e) {
      process.stderr.write(`[spec-serve] Error: ${e}
`);
    }
  });
}

// src/cli.ts
function normalizeDag(dag) {
  if (Array.isArray(dag)) {
    const v3Nodes = (dag[2] || []).map((n, i) => {
      const props = n[2] || [];
      const states = n[3] || [];
      const edges = n[4] || [];
      return [i, n[0], n[1] === "p" ? "prediction" : "entity", "", props, states, edges];
    });
    return { v: dag[0], p: dag[1], n: v3Nodes, tk: dag[3] };
  }
  return dag;
}
function resolvePath2(p) {
  if (p.startsWith("~"))
    p = join3(homedir2(), p.slice(1));
  const resolved = p.startsWith("/") ? p : join3(process.cwd(), p);
  if (!p.startsWith("/") && !p.startsWith("~")) {
    const rel = resolve2(process.cwd(), p);
    const cwd = process.cwd();
    const isDescendant = rel.startsWith(cwd + "/");
    const isSelf = rel === cwd;
    const isAncestor = cwd.startsWith(rel + "/");
    if (!isDescendant && !isSelf && !isAncestor) {
      throw new Error(`Path traversal blocked: ${p}`);
    }
    return rel;
  }
  return resolved;
}
function parseSections2(markdown) {
  const lines = markdown.split(`
`), sections = [];
  let heading = "(root)", level = 1, content = [], lineStart = 1;
  for (let i = 0;i < lines.length; i++) {
    const m2 = lines[i].match(/^##\s+(.+)/), m3 = lines[i].match(/^###\s+(.+)/);
    if (m2 && !m3) {
      if (content.length > 0 || sections.length === 0)
        sections.push({ heading, level, content: content.join(`
`).trim(), lineStart, lineEnd: i });
      heading = m2[1];
      level = 2;
      content = [];
      lineStart = i + 1;
    } else if (m3) {
      if (content.length > 0 || sections.length === 0)
        sections.push({ heading, level, content: content.join(`
`).trim(), lineStart, lineEnd: i });
      heading = m3[1];
      level = 3;
      content = [];
      lineStart = i + 1;
    } else {
      content.push(lines[i]);
    }
  }
  if (content.length > 0 || sections.length === 0)
    sections.push({ heading, level, content: content.join(`
`).trim(), lineStart, lineEnd: lines.length });
  return sections;
}
var program2 = new Command;
var pkg = JSON.parse(readFileSync3(new URL("../package.json", import.meta.url), "utf-8"));
program2.name("spec").alias("dotdog").description("CLI for structured software specs : validate .dog, compile .dag, query via MCP").version(pkg.version);
program2.command("validate [dir]").action((d = ".") => {
  const dir = resolvePath2(d);
  const dirs = [join3(dir, "projects"), join3(dir, "specs")];
  let found = false, hasErrors = false;
  for (const dd of dirs) {
    if (!existsSync3(dd))
      continue;
    const projects = readdirSync3(dd, { withFileTypes: true }).filter((e) => e.isDirectory()).map((e) => e.name);
    for (const p of projects) {
      found = true;
      const pd = join3(dd, p);
      if (!existsSync3(join3(pd, "SPEC.dog")))
        continue;
      const files = existsSync3(pd) ? readdirSync3(pd).filter((f) => f.endsWith(".dog")) : [];
      const missing = ["SPEC.dog", "constitution.dog", "data-model.dog"].filter((f) => !files.includes(f));
      const optional = ["COPY.dog", "plan.dog", "DESIGN-SYSTEM.dog", "INDEX.dog"].filter((f) => !files.includes(f));
      console.log(source_default.bold(`
  ${p} : ${files.length} .dog files, ${Math.max(0, 100 - Math.round(missing.length * 3 / 20 * 100))}% complete`));
      for (const f of files)
        console.log(source_default.gray(`    ${f}`));
      if (missing.length) {
        console.log(source_default.red(`  Missing required: ${missing.join(", ")}`));
        hasErrors = true;
      }
      if (optional.length)
        console.log(source_default.gray(`  Optional: ${optional.join(", ")} — not required for 100%`));
    }
  }
  if (!found)
    console.log(source_default.yellow("No projects found. Run: spec init <project>"));
  if (hasErrors)
    process.exit(1);
});
program2.command("init <project>").option("-m, --minimal", "Only SPEC.dog + data-model.dog").action((p, opts) => {
  const d = join3(process.cwd(), "specs", p);
  mkdirSync(d, { recursive: true });
  const full = {
    "SPEC.dog": `# Project

## Product

`,
    "constitution.dog": `# Constitution

1. **Rule.**
`,
    "data-model.dog": `# Data Model

## Entities

`,
    "plan.dog": `# Plan

## Phase 1

- [ ] Task
`,
    "COPY.dog": `# Copy

| Element | Copy |
|---|---|
`,
    "INDEX.dog": `# INDEX

| You | Start | Then |
|---|---|---|
`
  };
  const minimal = {
    "SPEC.dog": `# Project

## Product

`,
    "data-model.dog": `# Data Model

## Entities

`
  };
  const tmpl = opts.minimal ? minimal : full;
  for (const [f, c] of Object.entries(tmpl)) {
    writeFileSync2(join3(d, f), c);
    try {
      parse(c);
    } catch (_) {
      console.log(source_default.red(`  ✗ Template ${f} is invalid`));
      process.exit(1);
    }
    console.log(source_default.green(`  ✓ ${f}`));
  }
  console.log(source_default.bold(`
Project "${p}" initialized. Fill in SPEC.dog then run spec validate.`));
});
program2.command("list").option("--json", "Output project names as JSON").action((opts) => {
  const entries = [];
  for (const d of ["projects", "specs"]) {
    const dd = join3(process.cwd(), d);
    if (!existsSync3(dd))
      continue;
    const projects = readdirSync3(dd, { withFileTypes: true }).filter((e) => e.isDirectory()).map((e) => e.name);
    for (const p of projects) {
      const sp = join3(dd, p);
      const n = existsSync3(sp) ? readdirSync3(sp).filter((f) => f.endsWith(".dog")).length : 0;
      entries.push({ root: d, name: p, dogFiles: n });
    }
  }
  if (opts.json) {
    console.log(JSON.stringify(entries.map((e) => e.name)));
    return;
  }
  for (const d of ["projects", "specs"]) {
    const projects = entries.filter((e) => e.root === d);
    if (!projects.length)
      continue;
    console.log(source_default.bold(`
${d}/`));
    for (const p of projects) {
      console.log(`  ${source_default.cyan(p.name)} : ${p.dogFiles} .dog files`);
    }
  }
});
program2.command("parse <file>").action((f) => {
  const c = readFileSync3(f, "utf-8");
  const s = parseSections2(c);
  console.log(source_default.bold(`
${s.length} sections`));
  for (const sec of s)
    console.log(`  ${sec.heading.padEnd(30)} ${sec.content.length} chars`);
});
program2.command("compile [dir]").option("-o, --output <file>").option("--v3", "Ultra-compact v3 format (53% smaller than v2)").action((d = ".", opts) => {
  const dir = resolvePath2(d);
  const dirs = [join3(dir, "projects"), join3(dir, "specs"), dir];
  let found = false;
  for (const dd of dirs) {
    if (!existsSync3(dd))
      continue;
    const projects = readdirSync3(dd, { withFileTypes: true }).filter((e) => e.isDirectory()).map((e) => e.name);
    for (const p of projects) {
      const pd = join3(dd, p);
      if (!existsSync3(join3(pd, "SPEC.dog")))
        continue;
      if (!existsSync3(pd))
        continue;
      const files = readdirSync3(pd).filter((f) => f.endsWith(".dog")).sort();
      if (!files.length)
        continue;
      found = true;
      const sources = {};
      let sourceBytes = 0, contentBytes = 0;
      for (const f of files) {
        const content = readFileSync3(join3(pd, f), "utf-8");
        sources[f] = content;
        const bytes = Buffer.byteLength(content, "utf-8");
        sourceBytes += bytes;
        if (bytes >= 100)
          contentBytes += bytes;
      }
      const sourceTokens = Math.round(sourceBytes / 4);
      const contentTokens = Math.round(contentBytes / 4);
      const nodes = [], edges = [];
      for (const f of files) {
        const ast = parse(sources[f]);
        for (const section of ast.sections) {
          for (const block of section.blocks) {
            if (block.kind === "entity" || block.kind === "event") {
              const compactProps = {};
              for (const [key, val] of Object.entries(block.properties || {})) {
                let enc = "";
                const t = val.type || "string";
                if (t === "string")
                  enc = "s";
                else if (t === "number")
                  enc = "n";
                else if (t === "boolean")
                  enc = "b";
                else if (t === "enum")
                  enc = "e";
                else if (t === "json")
                  enc = "j";
                else
                  enc = t[0];
                if (val.required)
                  enc += "!";
                compactProps[key] = enc;
              }
              nodes.push({
                i: block.name || "",
                t: block.type || "",
                g: block.kind,
                d: block.description || "",
                p: compactProps,
                s: block.states || [],
                l: block.lifecycle || []
              });
            }
            if (block.kind === "prediction") {
              nodes.push({
                i: block.statement || block.name || "",
                t: "prediction",
                g: "prediction",
                d: block.description || "",
                p: {},
                s: [],
                l: [],
                cf: block.confidence || 0,
                tf: block.timeframe || "",
                tg: block.trigger || "",
                ms: block.measurement || ""
              });
            }
            if (block.kind === "relationship") {
              edges.push({
                s: block.source,
                t: block.target,
                v: block.verb,
                d: block.description || "",
                c: block.cardinality,
                r: block.required
              });
            }
          }
        }
      }
      const entityNames = new Set(nodes.map((n) => n.i));
      for (const e of edges) {
        if (e.s && !entityNames.has(e.s)) {
          console.log(source_default.red(`  ✗ Unknown relationship source "${e.s}" (target: "${e.t}")`));
          process.exit(1);
        }
        if (e.t && !entityNames.has(e.t)) {
          console.log(source_default.red(`  ✗ Unknown relationship target "${e.t}" (source: "${e.s}")`));
          process.exit(1);
        }
      }
      const nodeIds = new Map;
      nodes.forEach((n, i) => nodeIds.set(n.i, i));
      const v2nodes = [];
      for (let j = 0;j < nodes.length; j++) {
        const nd2 = nodes[j];
        const props = [];
        if (nd2.p)
          for (const [k, v] of Object.entries(nd2.p))
            props.push(k, v);
        const states = nd2.s || [];
        const outEdges = [];
        const seen = new Set;
        for (const e of edges) {
          if (e.s !== nd2.i && e.t !== nd2.i)
            continue;
          const tid = nodeIds.get(e.s === nd2.i ? e.t : e.s);
          if (tid === undefined)
            continue;
          const key = `${j}→${tid}:${e.v}`;
          if (seen.has(key))
            continue;
          seen.add(key);
          const ee = [tid, e.v || ""];
          if (e.c)
            ee.push(e.c);
          if (e.r)
            ee.push(1);
          outEdges.push(ee);
        }
        const entry = [j, nd2.i || "", nd2.t || "", nd2.d || "", props, states, outEdges];
        if (nd2.g === "prediction") {
          const f = [];
          if (nd2.cf)
            f.push(nd2.cf);
          if (nd2.tf)
            f.push(nd2.tf);
          if (nd2.tg)
            f.push(nd2.tg);
          if (nd2.ms)
            f.push(nd2.ms);
          if (f.length)
            entry.push(f);
        }
        v2nodes.push(entry);
      }
      const outPath = opts.output || join3(pd, `${p}.dag`);
      const v2dag = { v: 2, p, n: v2nodes };
      const v2Json = JSON.stringify(v2dag);
      const dagTokens = Math.round(Buffer.byteLength(v2Json, "utf-8") / 4);
      const allSavingsPct = sourceTokens > 0 ? Math.round((1 - dagTokens / sourceTokens) * 1000) / 10 : 0;
      const contentSavingsPct = contentTokens > 0 ? Math.round((1 - dagTokens / contentTokens) * 1000) / 10 : 0;
      const savingsTokens = sourceTokens - dagTokens;
      const tokens = { m: "chars/4", st: sourceTokens, ct: contentTokens, dt: dagTokens, sv: allSavingsPct, cs: contentSavingsPct, saved: savingsTokens };
      const v3dag = [3, p, v2nodes.map((n) => {
        const nd2 = nodes[n[0]];
        const tc = nd2.g === "prediction" ? "p" : "e";
        const st = nd2.s && nd2.s.length ? nd2.s : null;
        const ed = n[6] && n[6].length ? n[6] : null;
        const entry = [nd2.i, tc, n[4].length ? n[4] : null, st, ed];
        if (nd2.g === "prediction") {
          const f = [];
          if (nd2.cf != null)
            f.push(nd2.cf);
          if (nd2.tf)
            f.push(nd2.tf);
          if (f.length)
            entry.push(f);
        }
        return entry;
      }), tokens];
      const dag = opts.v3 ? v3dag : v2dag;
      const dagJson = JSON.stringify(dag);
      const report = opts.v3 ? [...dag, tokens] : { ...dag, tk: tokens };
      writeFileSync2(outPath, JSON.stringify(report, null, 2) + `
`);
      console.log(source_default.green(`  ✓ ${outPath}`));
      console.log(source_default.gray(`    ${nodes.length} nodes, ${edges.length} edges, ${files.length} files`));
      console.log(source_default.gray(`    ${sourceTokens} → ${dagTokens} tokens (${allSavingsPct}% savings, ${contentSavingsPct}% content-only, ${savingsTokens} saved)`));
    }
  }
  if (!found)
    console.log(source_default.yellow("No projects found."));
});
program2.command("tokens [dir]").action((d = ".") => {
  const dir = resolvePath2(d);
  const dirs = [join3(dir, "projects"), join3(dir, "specs"), dir];
  let found = false;
  for (const dd of dirs) {
    if (!existsSync3(dd))
      continue;
    const projects = readdirSync3(dd, { withFileTypes: true }).filter((e) => e.isDirectory()).map((e) => e.name);
    for (const p of projects) {
      const pd = join3(dd, p);
      if (!existsSync3(join3(pd, "SPEC.dog")))
        continue;
      const dagFile = join3(pd, `${p}.dag`);
      if (!existsSync3(dagFile))
        continue;
      found = true;
      const dogFiles = readdirSync3(pd).filter((f) => f.endsWith(".dog"));
      let sourceBytes = 0, contentBytes = 0;
      for (const f of dogFiles) {
        const bytes = Buffer.byteLength(readFileSync3(join3(pd, f), "utf-8"), "utf-8");
        sourceBytes += bytes;
        if (bytes >= 100)
          contentBytes += bytes;
      }
      const dagBytes = Buffer.byteLength(readFileSync3(dagFile, "utf-8"), "utf-8");
      const savings = sourceBytes > 0 ? Math.round((1 - dagBytes / sourceBytes) * 1000) / 10 : 0;
      const dag = JSON.parse(readFileSync3(dagFile, "utf-8"));
      const dagOnly = JSON.stringify({ v: dag.v, p: dag.p, n: dag.n });
      const dagOnlyBytes = Buffer.byteLength(dagOnly, "utf-8");
      const dagOnlyPct = sourceBytes > 0 ? Math.round((1 - dagOnlyBytes / sourceBytes) * 1000) / 10 : 0;
      console.log(source_default.bold(`
  ${p}`));
      console.log(source_default.gray(`    ${dogFiles.length} .dog files: ${sourceBytes} bytes`));
      console.log(source_default.gray(`    .dag on disk: ${dagBytes} bytes (${savings}% savings, includes metadata)`));
      console.log(source_default.green(`    .dag payload: ${dagOnlyBytes} bytes (${dagOnlyPct}% savings, graph only)`));
      if (contentBytes && contentBytes !== sourceBytes) {
        const cs = Math.round((1 - dagOnlyBytes / contentBytes) * 1000) / 10;
        console.log(source_default.gray(`    content-only: ${contentBytes} bytes → ${cs}% savings`));
      }
    }
  }
  if (!found)
    console.log(source_default.yellow("No .dag files found. Run compile first."));
});
program2.command("visualize [dir]").option("-s, --save").action((d = ".", opts) => {
  const dir = resolvePath2(d);
  const dirs = [join3(dir, "projects"), join3(dir, "specs"), dir];
  for (const dd of dirs) {
    if (!existsSync3(dd))
      continue;
    const projects = readdirSync3(dd, { withFileTypes: true }).filter((e) => e.isDirectory()).map((e) => e.name);
    for (const p of projects) {
      const dagFile = join3(dd, p, `${p}.dag`);
      if (!existsSync3(dagFile))
        continue;
      const dag = JSON.parse(readFileSync3(dagFile, "utf-8"));
      const d2 = normalizeDag(dag);
      const nodes = d2.n || dag.nodes || [];
      const isV22 = (n) => Array.isArray(n) && typeof n[0] === "number";
      const nodeName2 = (n) => isV22(n) ? nodes[n[0]] ? nodes[n[0]][1] || String(n[0]) : String(n[0]) : n.i || n.id || "";
      const slug = (s) => s.replace(/\s+/g, "_").replace(/^[^a-zA-Z]+/, "n_");
      let out = "```mermaid\ngraph LR\n";
      for (const n of nodes) {
        const raw = isV22(n) ? n[1] || String(n[0]) : n.i || n.id || "";
        if (raw)
          out += `    ${slug(raw)}[${raw}]
`;
      }
      const seen = new Set;
      for (const n of nodes) {
        const edges = isV22(n) ? n[6] || [] : n.es || [];
        for (const e of edges) {
          const srcName = isV22(n) ? n[1] || String(n[0]) : n.i || n.id || "";
          const src = slug(srcName);
          let tgtName, verb;
          if (isV22(n)) {
            const tgtNode = nodes[e[0]];
            tgtName = tgtNode ? tgtNode[1] || String(tgtNode[0]) : String(e[0]);
            verb = e[1] || "";
          } else {
            tgtName = e.t || "";
            verb = e.v || "";
          }
          const tgt = slug(tgtName);
          const key = `${src}→${tgt}:${verb}`;
          if (!seen.has(key) && src && tgt) {
            seen.add(key);
            out += `    ${src} -->|${verb}| ${tgt}
`;
          }
        }
      }
      const legacyEdges = dag.e || dag.edges || [];
      for (const e of legacyEdges) {
        const src = slug(e.s || e.source || "");
        const tgt = slug(e.t || e.target || "");
        const verb = e.v || e.verb || "";
        const key = `${src}→${tgt}:${verb}`;
        if (!seen.has(key) && src && tgt) {
          seen.add(key);
          out += `    ${src} -->|${verb}| ${tgt}
`;
        }
      }
      out += "```\n";
      if (opts.save) {
        const outFile = join3(dd, p, `${p}.md`);
        writeFileSync2(outFile, `# ${p} : Spec Graph

${out}`);
        console.log(source_default.green(`  ✓ ${outFile}`));
      }
      console.log(out);
    }
  }
});
program2.command("serve [dir]").description("MCP server : expose .dag graph to AI agents over stdio").action((d = ".") => serve(resolvePath2(d)));
program2.command("analyze [dir]").description("Analyze a spec project : score, gaps, suggestions").option("-p, --project <name>").action((d = ".", opts) => {
  const dir = resolvePath2(d);
  const dirs = [join3(dir, "projects"), join3(dir, "specs"), dir];
  console.log(source_default.bold(`
Spec Analysis
`));
  let found = false, hasGaps = false;
  for (const dd of dirs) {
    if (!existsSync3(dd))
      continue;
    const projects = readdirSync3(dd, { withFileTypes: true }).filter((e) => e.isDirectory()).map((e) => e.name);
    for (const p of projects) {
      if (opts.project && p !== opts.project)
        continue;
      const pd = join3(dd, p);
      if (!existsSync3(join3(pd, "SPEC.dog")))
        continue;
      if (!existsSync3(pd))
        continue;
      const files = readdirSync3(pd).filter((f) => f.endsWith(".dog"));
      if (!files.length)
        continue;
      found = true;
      console.log(source_default.bold(`
  ${p}`));
      console.log("  " + "─".repeat(50));
      const allEntities = [];
      const allRelationships = [];
      const analyses = [];
      for (const f of files) {
        const content = readFileSync3(join3(pd, f), "utf-8");
        const ast = parse(content);
        const entities = ast.sections.flatMap((s) => s.blocks.filter((b) => b.kind === "entity"));
        const rels = ast.sections.flatMap((s) => s.blocks.filter((b) => b.kind === "relationship"));
        allEntities.push(...entities);
        allRelationships.push(...rels);
        analyses.push({ file: f, sections: ast.sections.length, size: content.length, entities: entities.length, rels: rels.length });
      }
      const missingReq = ["SPEC.dog", "constitution.dog", "data-model.dog"].filter((f) => !files.includes(f));
      const missingOpt = ["COPY.dog", "plan.dog", "DESIGN-SYSTEM.dog", "INDEX.dog"].filter((f) => !files.includes(f));
      let score = 100 - missingReq.length * 15 - missingOpt.length * 5;
      const noDesc = allEntities.filter((e) => !e.description || e.description.length < 10).length;
      score = Math.max(0, score - noDesc * 3);
      const noProps = allEntities.filter((e) => Object.keys(e.properties).length === 0).length;
      score = Math.max(0, score - noProps * 5);
      const noStates = allEntities.filter((e) => e.states.length === 0).length;
      score = Math.max(0, score - noStates * 3);
      console.log(`  ${files.length} files | ${score}% complete`);
      for (const a of analyses) {
        const detail = a.entities > 0 ? ` (${a.entities} entities, ${a.rels} rels)` : "";
        console.log(source_default.gray(`    ${a.file} : ${a.sections} sections, ${(a.size / 1024).toFixed(1)}KB${detail}`));
      }
      const gaps = [];
      for (const f of missingReq)
        gaps.push(`\uD83D\uDD34 ${f}: Missing required file`);
      for (const f of missingOpt)
        gaps.push(`ℹ️ ${f}: Optional file not present`);
      const entityNames = new Set(allEntities.map((e) => e.name));
      for (const e of allEntities) {
        if (!e.description || e.description.length < 10)
          gaps.push(`\uD83D\uDFE1 ${e.name}: No description`);
        if (Object.keys(e.properties).length === 0)
          gaps.push(`\uD83D\uDFE1 ${e.name}: No properties defined`);
        if (e.states.length === 0)
          gaps.push(`\uD83D\uDD35 ${e.name}: No states defined`);
      }
      for (const r of allRelationships) {
        if (r.source && !entityNames.has(r.source))
          gaps.push(`\uD83D\uDFE1 Relationship: unknown source "${r.source}"`);
        if (r.target && !entityNames.has(r.target))
          gaps.push(`\uD83D\uDFE1 Relationship: unknown target "${r.target}"`);
      }
      const contradictions = [];
      const relMap = new Map;
      for (const r of allRelationships) {
        const key = `${r.source}→${r.target}`;
        if (!relMap.has(key))
          relMap.set(key, []);
        relMap.get(key).push(r);
      }
      for (const [key, rels] of relMap) {
        if (rels.length > 1) {
          const verbs = [...new Set(rels.map((r) => r.verb))];
          if (verbs.length > 1)
            contradictions.push(`\uD83D\uDD34 Contradiction: "${key}" has conflicting verbs: ${verbs.join(", ")}`);
          else
            contradictions.push(`\uD83D\uDFE1 Duplicate: "${key}" appears ${rels.length} times with same verb "${verbs[0]}"`);
        }
      }
      for (const r of allRelationships) {
        const reverse = allRelationships.find((r2) => r2.source === r.target && r2.target === r.source);
        if (reverse && r.verb !== reverse.verb && r.source < r.target) {
          contradictions.push(`\uD83D\uDFE1 Bidirectional: "${r.source}→${r.target}" verb "${r.verb}" vs "${reverse.source}→${reverse.target}" verb "${reverse.verb}"`);
        }
      }
      for (const c of contradictions) {
        if (c.startsWith("\uD83D\uDD34"))
          hasGaps = true;
      }
      if (contradictions.length > 0) {
        console.log(source_default.bold(`
  Contradictions (${contradictions.length})`));
        for (const c of contradictions)
          console.log(`  ${c}`);
      }
      if (gaps.length > 0) {
        console.log(source_default.bold(`
  Gaps (${gaps.length})`));
        for (const g of gaps)
          console.log(`  ${g}`);
      } else
        console.log(source_default.green(`
  No gaps found.`));
      if (gaps.length > 0)
        hasGaps = true;
    }
  }
  if (!found)
    console.log(source_default.yellow("No spec projects found. Run: dotdog init <project>"));
  if (hasGaps)
    process.exit(1);
});
program2.command("generate [dir]").description("Generate missing spec files from SPEC.dog").option("-p, --project <name>").action((d = ".", opts) => {
  const dir = resolvePath2(d);
  const dirs = [join3(dir, "projects"), join3(dir, "specs"), dir];
  let specContent = "", specDir = "";
  for (const dd of dirs) {
    if (!existsSync3(dd))
      continue;
    const projects = readdirSync3(dd, { withFileTypes: true }).filter((e) => e.isDirectory()).map((e) => e.name);
    for (const p of projects) {
      if (opts.project && p !== opts.project)
        continue;
      const pd = join3(dd, p);
      const sp = join3(pd, "SPEC.dog");
      if (existsSync3(sp)) {
        specContent = readFileSync3(sp, "utf-8");
        specDir = pd;
        break;
      }
    }
    if (specContent)
      break;
  }
  if (!specContent) {
    console.log(source_default.red("No SPEC.dog found. Create one first."));
    return;
  }
  console.log(source_default.bold(`
Spec Generator
`));
  console.log(source_default.gray(`  Source: ${specDir}/SPEC.dog
`));
  const ast = parse(specContent);
  const entities = ast.sections.flatMap((s) => s.blocks.filter((b) => b.kind === "entity"));
  const uiStrings = [];
  for (const section of ast.sections) {
    const h = section.heading.toLowerCase();
    if (h.includes("what the user sees") || h.includes("screen")) {
      const text = section.blocks.filter((b) => b.kind === "prose").map((b) => b.content).join(`
`);
      for (const m of text.match(/\[([^\]]+)\]/g) || [])
        uiStrings.push({ screen: section.heading, element: "button", text: m });
      for (const m of text.match(/"([^"]+)"/g) || [])
        uiStrings.push({ screen: section.heading, element: "label", text: m });
    }
  }
  if (!existsSync3(join3(specDir, "data-model.dog")) && entities.length > 0) {
    let dm = `# Data Model

## Core Entities

`;
    for (const e of entities) {
      dm += `### Entity: ${e.name}

${e.description || "No description."}

`;
      dm += "```yaml\n";
      dm += `entity: ${e.name}
`;
      dm += `type: entity
`;
      dm += `properties:
`;
      for (const [k, v] of Object.entries(e.properties)) {
        dm += `  ${k}:
`;
        dm += `    type: ${v.type}
`;
        if (v.required)
          dm += `    required: true
`;
      }
      if (e.states.length > 0)
        dm += `states: [${e.states.join(", ")}]
`;
      dm += "```\n\n";
    }
    writeFileSync2(join3(specDir, "data-model.dog"), dm);
    try {
      parse(dm);
    } catch (_) {
      console.log(source_default.red("  ✗ Generated data-model.dog is invalid"));
      process.exit(1);
    }
    console.log(source_default.green(`  ✓ data-model.dog (${entities.length} entities)`));
  }
  if (!existsSync3(join3(specDir, "COPY.dog")) && uiStrings.length > 0) {
    let copy = `# App Copy

| Screen | Element | Copy |
|--------|---------|------|
`;
    for (const s of uiStrings)
      copy += `| ${s.screen} | ${s.element} | ${s.text} |
`;
    writeFileSync2(join3(specDir, "COPY.dog"), copy);
    try {
      parse(copy);
    } catch (_) {
      console.log(source_default.red("  ✗ Generated COPY.dog is invalid"));
      process.exit(1);
    }
    console.log(source_default.green(`  ✓ COPY.dog (${uiStrings.length} strings)`));
  }
  if (!existsSync3(join3(specDir, "INDEX.dog"))) {
    let idx = `# INDEX

| You are... | Start here | Then... |
|------------|-----------|---------|
`;
    idx += `| Developer | SPEC.dog | data-model.dog → plan.dog |
`;
    idx += `| AI agent | data-model.dog | COPY.dog → SPEC.dog |
`;
    idx += `| Designer | SPEC.dog | COPY.dog |
`;
    writeFileSync2(join3(specDir, "INDEX.dog"), idx);
    try {
      parse(idx);
    } catch (_) {
      console.log(source_default.red("  ✗ Generated INDEX.dog is invalid"));
      process.exit(1);
    }
    console.log(source_default.green("  ✓ INDEX.dog"));
  }
  console.log(source_default.bold(`
Run dotdog validate to verify.
`));
});
program2.command("simulate <scenario>").description("Walk through a scenario, check pre/postconditions").option("-p, --project <name>").action((scenario, opts) => {
  const dir = resolvePath2(".");
  const dirs = [join3(dir, "projects"), join3(dir, "specs"), dir];
  let projectDir = "";
  for (const dd of dirs) {
    if (!existsSync3(dd))
      continue;
    const projects = readdirSync3(dd, { withFileTypes: true }).filter((e) => e.isDirectory()).map((e) => e.name);
    for (const p of projects) {
      if (opts.project && p !== opts.project)
        continue;
      const pd = join3(dd, p);
      if (existsSync3(join3(pd, "SPEC.dog"))) {
        projectDir = pd;
        break;
      }
    }
    if (projectDir)
      break;
  }
  if (!projectDir) {
    console.log(source_default.red("Project not found."));
    return;
  }
  const dagFile = join3(projectDir, `${opts.project || projectDir.split("/").pop()}.dag`);
  let entities = [], relationships = [];
  if (existsSync3(dagFile)) {
    const dag = JSON.parse(readFileSync3(dagFile, "utf-8"));
    const d2 = normalizeDag(dag);
    const simNodes = d2.n || dag.nodes || [];
    const isV22 = (n) => Array.isArray(n) && typeof n[0] === "number";
    entities = simNodes.map((n) => (isV22(n) ? n[1] || String(n[0]) : n.i || n.id || "").toLowerCase());
    if (dag.e || dag.edges) {
      relationships = dag.e || dag.edges || [];
    } else {
      const seen = new Set;
      for (const n of simNodes) {
        const edges = isV22(n) ? n[6] || [] : n.es || [];
        for (const e of edges) {
          const srcName = isV22(n) ? n[1] || String(n[0]) : n.i || n.id || "";
          const tgtName = isV22(n) ? (simNodes[e[0]] ? simNodes[e[0]][2] : "") || String(e[0]) : e.t || "";
          const verb = isV22(n) ? e[1] || "" : e.v || "";
          const key = `${srcName}→${tgtName}:${verb}`;
          if (!seen.has(key)) {
            seen.add(key);
            relationships.push({ s: srcName, t: tgtName, v: verb });
          }
        }
      }
    }
  }
  const specFile = join3(projectDir, "SPEC.dog");
  const specContent = existsSync3(specFile) ? readFileSync3(specFile, "utf-8") : "";
  console.log(source_default.bold(`
Simulation: ${scenario}`));
  console.log(source_default.gray(`Project: ${projectDir.split("/").pop()} | Entities: ${entities.length} | Relationships: ${relationships.length}`));
  const steps = [];
  const stepMatches = specContent.match(/\[(\d+)\/\d+\]\s*(.+)/g);
  if (stepMatches) {
    for (const m of stepMatches) {
      const step = m.replace(/\[\d+\/\d+\]\s*/, "");
      if (step)
        steps.push(step);
    }
  }
  if (steps.length === 0) {
    console.log(source_default.yellow(`
  No scenario steps found in SPEC.dog.`));
    console.log(source_default.gray("  Add steps like: [1/3] User taps button"));
    return;
  }
  let passed = 0, failed = 0;
  for (let i = 0;i < steps.length; i++) {
    const step = steps[i];
    console.log(source_default.cyan(`
  [${i + 1}/${steps.length}] ${step}`));
    let foundRef = false;
    for (const e of entities) {
      if (step.toLowerCase().includes(e)) {
        console.log(source_default.green(`    ✓ References entity: ${e}`));
        foundRef = true;
      }
    }
    for (const r of relationships) {
      const src = (r.s || r.source || "").toLowerCase();
      const tgt = (r.t || r.target || "").toLowerCase();
      const verb = (r.v || r.verb || "").toLowerCase();
      if (step.toLowerCase().includes(src) && step.toLowerCase().includes(tgt)) {
        console.log(source_default.green(`    ✓ References relationship: ${src} → ${tgt}`));
        foundRef = true;
      }
      if (verb && step.toLowerCase().includes(verb)) {
        console.log(source_default.green(`    ✓ References verb: ${verb}`));
        foundRef = true;
      }
    }
    if (!foundRef) {
      console.log(source_default.yellow(`    ⚠ No entity or relationship references found`));
      failed++;
    } else {
      passed++;
    }
  }
  console.log(source_default.bold(`
  RESULT: ${failed === 0 ? source_default.green("Success") : source_default.yellow("Partial")} (${passed}/${steps.length} steps passed)`));
  if (failed > 0)
    console.log(source_default.red(`  ${failed} steps have no spec references — entities or relationships may be missing.`));
  if (failed === 0)
    console.log(source_default.green("  All steps reference known entities and relationships."));
});
program2.command("staleness [dir]").action((d = ".") => {
  const dir = resolvePath2(d);
  const dirs = [join3(dir, "projects"), join3(dir, "specs"), dir];
  console.log(source_default.bold(`Staleness Audit
`));
  for (const dd of dirs) {
    if (!existsSync3(dd))
      continue;
    const projects = readdirSync3(dd, { withFileTypes: true }).filter((e) => e.isDirectory()).map((e) => e.name);
    for (const p of projects) {
      const pd = join3(dd, p);
      if (!existsSync3(join3(pd, "SPEC.dog")))
        continue;
      if (!existsSync3(pd))
        continue;
      const planFile = join3(pd, "plan.dog");
      if (!existsSync3(planFile)) {
        console.log(source_default.yellow(`  ${p}: No plan.dog`));
        continue;
      }
      const plan = readFileSync3(planFile, "utf-8");
      const tasks = [...plan.matchAll(/^\s*- \[([ x])\]\s+(.+)/gm)];
      let issues = 0;
      for (const m of tasks) {
        const done = m[1] === "x";
        const text = m[2].toLowerCase();
        const precedingText = plan.substring(Math.max(0, m.index - 200), m.index);
        const phaseMatch = precedingText.match(/Phase\s+(\d+)/);
        const phase = phaseMatch ? parseInt(phaseMatch[1]) : 99;
        if (phase > 3)
          continue;
        if ((text.includes("npm publish") || text.includes("npm install")) && !done) {
          const pkgPath = join3(resolvePath2("."), "packages/dotdog/package.json");
          if (existsSync3(pkgPath)) {
            const pkg2 = JSON.parse(readFileSync3(pkgPath, "utf-8"));
            if (pkg2.version) {
              console.log(source_default.yellow(`  ⚠ Should be [x]: ${m[2].trim()}`));
              issues++;
            }
          }
        }
        if (text.includes("compile")) {
          if (!done) {
            console.log(source_default.yellow(`  ⚠ Should be [x]: ${m[2].trim()}`));
            issues++;
          }
        }
        if (text.includes("generate") && !done) {
          if (existsSync3(join3(dir, "packages/dotdog/src/cli.ts")) || existsSync3(join3(dir, "packages/spec-cli/src/generate.ts"))) {
            console.log(source_default.yellow(`  ⚠ Should be [x]: ${m[2].trim()}`));
            issues++;
          }
        }
      }
      if (issues === 0)
        console.log(source_default.green(`  ${p}: spec matches reality`));
      else
        console.log(source_default.bold(`  ${issues} stale items. Update plan.dog.`));
    }
  }
});
program2.command("verify [dir]").description("Verify spec-code alignment. --init auto-generates verify section in plan.dog").option("-i, --init", "Auto-generate verify section from codebase scan").action((d = ".", opts) => {
  const dir = resolvePath2(d);
  const dirs = [join3(dir, "projects"), join3(dir, "specs"), dir];
  console.log(source_default.bold(opts.init ? `Auto-Generating Verify Section
` : `Verification Audit
`));
  for (const dd of dirs) {
    if (!existsSync3(dd))
      continue;
    const projects = readdirSync3(dd, { withFileTypes: true }).filter((e) => e.isDirectory()).map((e) => e.name);
    for (const p of projects) {
      const pd = join3(dd, p);
      if (!existsSync3(join3(pd, "SPEC.dog")))
        continue;
      const planFile = join3(pd, "plan.dog");
      if (!existsSync3(planFile)) {
        console.log(source_default.yellow(`  ${p}: No plan.dog`));
        continue;
      }
      const dagFile = join3(pd, `${p}.dag`);
      if (!existsSync3(dagFile)) {
        console.log(source_default.yellow(`  ${p}: No .dag file. Run compile first.`));
        continue;
      }
      const dag = JSON.parse(readFileSync3(dagFile, "utf-8"));
      const plan = readFileSync3(planFile, "utf-8");
      if (opts.init) {
        const entities = [];
        const props = new Map;
        const d3 = normalizeDag(dag);
        for (const node of d3.n || []) {
          const name = node[1] || String(node[0]);
          entities.push(name);
          if (node[4])
            props.set(name, node[4].map((p2) => p2[0]));
        }
        let verify = `
## Verify

`;
        for (const entity of entities) {
          const nameLower = entity.toLowerCase().replace(/[^a-z0-9]/g, "");
          let matchFile = "";
          const skip = new Set(["node_modules", ".git", "dist", ".bun", "dev", "build"]);
          const codeDirs = [join3(dir, "src"), join3(dir, "lib"), join3(dir, "app"), dir];
          for (const cd of codeDirs) {
            if (!existsSync3(cd))
              continue;
            try {
              const allFiles = readdirSync3(cd, { recursive: true }).filter((f) => {
                if (!(f.endsWith(".ts") || f.endsWith(".js") || f.endsWith(".py") || f.endsWith(".sol") || f.endsWith(".go")))
                  return false;
                for (const part of f.split("/")) {
                  if (skip.has(part))
                    return false;
                }
                return true;
              });
              const match = allFiles.find((f) => f.toLowerCase().includes(nameLower));
              if (match) {
                matchFile = join3(cd, match).replace(dir, ".");
                break;
              }
            } catch (_) {}
          }
          verify += `### Entity: ${entity}
`;
          if (matchFile) {
            verify += `  file: ${matchFile}
`;
            const fullPath = join3(dir, matchFile.replace("./", ""));
            if (existsSync3(fullPath)) {
              const code = readFileSync3(fullPath, "utf-8");
              const codeProps = [...code.matchAll(/\b(\w+)\s*[:?]\s*\w+/g)].map((m) => m[1]).filter((v, i, a) => a.indexOf(v) === i);
              if (codeProps.length > 0) {
                verify += `  properties: [${codeProps.slice(0, 10).join(", ")}]
`;
              }
            }
          } else {
            verify += `  # no matching file found — map manually
`;
          }
          verify += `
`;
        }
        const updatedPlan = plan.includes("## Verify") ? plan : plan + verify;
        writeFileSync2(planFile, updatedPlan);
        console.log(source_default.green(`  ${p}: Verify section generated in plan.dog`));
      } else {
        const verifyMatch = plan.match(/## Verify\n([\s\S]*?)(?=\n## |$)/);
        if (!verifyMatch) {
          console.log(source_default.yellow(`  ${p}: No ## Verify section. Run: dotdog verify --init`));
          continue;
        }
        const verifyBlock = verifyMatch[1];
        const entityBlocks = [...verifyBlock.matchAll(/### Entity: (\w+)\n([\s\S]*?)(?=### Entity:|$)/g)];
        let checks = 0, passed = 0;
        for (const [, ename, ebody] of entityBlocks) {
          checks++;
          const fileMatch = ebody.match(/file:\s*(.+)/);
          const propMatch = ebody.match(/properties:\s*\[([^\]]+)\]/);
          if (!fileMatch)
            continue;
          const filePath = join3(dir, fileMatch[1].trim().replace("./", ""));
          if (!existsSync3(filePath)) {
            console.log(source_default.red(`  ✗ ${ename}: file ${fileMatch[1].trim()} not found`));
            continue;
          }
          const code = readFileSync3(filePath, "utf-8");
          const cleanCode = code.replace(/\/\/.*/g, "").replace(/\/\*[\s\S]*?\*\//g, "");
          if (propMatch) {
            const props = propMatch[1].split(",").map((s) => s.trim());
            let propPass = 0;
            for (const prop of props) {
              const snakeVariant = prop.replace(/_/g, "");
              const camelVariant = prop.replace(/_([a-z])/g, (_, c) => c.toUpperCase());
              if (cleanCode.includes(prop) || cleanCode.includes(snakeVariant) || cleanCode.includes(camelVariant)) {
                propPass++;
              } else {
                console.log(source_default.yellow(`  ⚠ ${ename}.${prop}: not found in ${fileMatch[1].trim()}`));
              }
            }
            if (propPass === props.length)
              passed++;
          }
        }
        if (checks === 0)
          console.log(source_default.yellow(`  ${p}: No entities mapped. Run: dotdog verify --init`));
        else if (passed === checks)
          console.log(source_default.green(`  ${p}: ${passed}/${checks} entities verified`));
        else
          console.log(source_default.bold(`  ${p}: ${passed}/${checks} entities verified`));
      }
    }
  }
});
program2.command("woof").action(() => {
  console.log("  / \\__");
  console.log(" (    @\\___");
  console.log("  /       O");
  console.log(" /   (_____/");
  console.log("/_____/   U");
});
program2.command("index [dir]").description("Build search index for semantic queries").action((d = ".") => {
  const dir = resolvePath2(d);
  const dirs = [join3(dir, "projects"), join3(dir, "specs"), dir];
  let built = 0;
  for (const dd of dirs) {
    if (!existsSync3(dd))
      continue;
    const projects = readdirSync3(dd, { withFileTypes: true }).filter((e) => e.isDirectory()).map((e) => e.name);
    for (const p of projects) {
      const pd = join3(dd, p);
      if (!existsSync3(join3(pd, "SPEC.dog")))
        continue;
      const idx = buildIndex(pd, p);
      writeFileSync2(join3(pd, `${p}.idx`), JSON.stringify(idx));
      console.log(source_default.green(`  ✓ ${p} : ${idx.entries.length} sections indexed (${idx.vocabulary.length} terms)`));
      built++;
    }
  }
  if (!built)
    console.log(source_default.yellow("No projects found. Run dotdog init first."));
});
program2.command("search <query>").description("Semantic search across compiled specs").option("-p, --project <name>").action((query, opts) => {
  console.log(source_default.bold(`
Search: "${query}"
`));
  const dir = resolvePath2(".");
  const dirs = [join3(dir, "projects"), join3(dir, "specs"), dir];
  let found = false;
  for (const dd of dirs) {
    if (!existsSync3(dd))
      continue;
    const projects = readdirSync3(dd, { withFileTypes: true }).filter((e) => e.isDirectory()).map((e) => e.name);
    for (const p of projects) {
      if (opts.project && p !== opts.project)
        continue;
      const pd = join3(dd, p);
      const idxFile = join3(pd, `${p}.idx`);
      if (!existsSync3(idxFile))
        continue;
      found = true;
      const idx = JSON.parse(readFileSync3(idxFile, "utf-8"));
      const results = searchIndex(idx, query, 8);
      if (results.length === 0) {
        console.log(source_default.gray(`  ${p}: No matches`));
        continue;
      }
      console.log(source_default.green(`  ${p} — ${results.length} results:`));
      for (const r of results) {
        const preview = r.entry.content.replace(/\n/g, " ").slice(0, 100);
        console.log(source_default.gray(`    ${Math.round(r.score * 100)}%  [${r.entry.file}] ${r.entry.heading}`));
        console.log(source_default.gray(`         ${preview}...`));
      }
    }
  }
  if (!found)
    console.log(source_default.yellow("No index found. Run dotdog index first."));
});
program2.command("predictions [dir]").description("List all predictions with status").option("-p, --project <name>").action((d = ".", opts) => {
  const dir = resolvePath2(d);
  const dirs = [join3(dir, "projects"), join3(dir, "specs"), dir];
  console.log(source_default.bold(`
Predictions
`));
  let found = false;
  for (const dd of dirs) {
    if (!existsSync3(dd))
      continue;
    const projects = readdirSync3(dd, { withFileTypes: true }).filter((e) => e.isDirectory()).map((e) => e.name);
    for (const p of projects) {
      if (opts.project && p !== opts.project)
        continue;
      const pd = join3(dd, p);
      if (!existsSync3(join3(pd, "SPEC.dog")))
        continue;
      const files = readdirSync3(pd).filter((f) => f.endsWith(".dog"));
      for (const f of files) {
        const content = readFileSync3(join3(pd, f), "utf-8");
        const ast = parse(content);
        for (const section of ast.sections) {
          for (const block of section.blocks) {
            if (block.kind === "prediction") {
              found = true;
              const b = block;
              const status = b.status || "pending";
              const icon = status === "correct" ? "✅" : status === "wrong" ? "❌" : status === "partial" ? "⚠️" : "⏳";
              console.log(`  ${icon} ${b.statement || b.name} (${(b.confidence * 100).toFixed(0)}% confidence, ${status})`);
              if (b.measurement)
                console.log(source_default.gray(`    Measurement: ${b.measurement}`));
              if (b.timeframe)
                console.log(source_default.gray(`    Timeframe: ${b.timeframe}`));
            }
          }
        }
      }
    }
  }
  if (!found)
    console.log(source_default.yellow("No predictions found."));
});
program2.command("resolve <name>").description("Mark a prediction as correct, wrong, or partial").option("-p, --project <name>").option("--correct", "Prediction was correct").option("--wrong", "Prediction was wrong").option("--partial", "Prediction was partially correct").action((name, opts) => {
  const dir = resolvePath2(".");
  const dirs = [join3(dir, "projects"), join3(dir, "specs"), dir];
  const status = opts.correct ? "correct" : opts.wrong ? "wrong" : opts.partial ? "partial" : null;
  if (!status) {
    console.log(source_default.red("Specify --correct, --wrong, or --partial"));
    return;
  }
  for (const dd of dirs) {
    if (!existsSync3(dd))
      continue;
    const projects = readdirSync3(dd, { withFileTypes: true }).filter((e) => e.isDirectory()).map((e) => e.name);
    for (const p of projects) {
      if (opts.project && p !== opts.project)
        continue;
      const pd = join3(dd, p);
      if (!existsSync3(join3(pd, "SPEC.dog")))
        continue;
      const files = readdirSync3(pd).filter((f) => f.endsWith(".dog"));
      for (const f of files) {
        const fp = join3(pd, f);
        let content = readFileSync3(fp, "utf-8");
        const originalContent = readFileSync3(fp, "utf-8");
        const ast = parse(content);
        for (const section of ast.sections) {
          for (const block of section.blocks) {
            if (block.kind === "prediction") {
              const b = block;
              if ((b.statement || b.name || "").toLowerCase().includes(name.toLowerCase())) {
                let headingIdx = -1;
                const stmt = b.statement || b.name || "";
                for (const prefix of ["###", "####", "#####"]) {
                  for (const fmt of [`${prefix} Prediction: ${stmt}`, `${prefix} ${stmt}`]) {
                    const idx = content.indexOf(fmt);
                    if (idx >= 0) {
                      headingIdx = idx;
                      break;
                    }
                  }
                  if (headingIdx >= 0)
                    break;
                }
                if (headingIdx >= 0) {
                  const blockStart = content.indexOf("```yaml", headingIdx);
                  const blockEnd = content.indexOf("```", blockStart + 7);
                  if (blockStart >= 0 && blockEnd >= 0) {
                    const yamlBlock = content.slice(blockStart, blockEnd + 3);
                    let newYaml = yamlBlock;
                    if (yamlBlock.includes("status:")) {
                      newYaml = yamlBlock.replace(/status:\s*\w+/, `status: ${status}`);
                    } else {
                      newYaml = yamlBlock.replace(/\n\s*(trigger|timeframe|confidence|measurement):/, `
  status: ${status}
$&`);
                    }
                    content = content.replace(yamlBlock, newYaml);
                    try {
                      parse(content);
                      writeFileSync2(fp, content, "utf-8");
                      console.log(source_default.green(`  ✓ ${b.statement || b.name}: ${status}`));
                    } catch (_) {
                      writeFileSync2(fp, originalContent, "utf-8");
                      console.log(source_default.red(`  ✗ resolve produced invalid output for "${name}" — reverted`));
                      process.exit(1);
                    }
                    return;
                  }
                }
              }
            }
          }
        }
      }
    }
  }
  console.log(source_default.yellow(`Prediction "${name}" not found.`));
});
var kitDir = join3(resolve2(import.meta.dirname || "."), "..", "kits");
var builtInKits = existsSync3(kitDir) ? readdirSync3(kitDir, { withFileTypes: true }).filter((e) => e.isDirectory()).map((e) => e.name) : [];
var kitCmd = program2.command("kit").description("Manage spec kits");
kitCmd.command("list").description("List available kits").action(() => {
  console.log(source_default.bold(`
Available kits
`));
  if (builtInKits.length === 0) {
    console.log(source_default.gray("  No built-in kits found."));
  } else {
    for (const k of builtInKits) {
      const specFile = join3(kitDir, k, "SPEC.dog");
      const desc = existsSync3(specFile) ? readFileSync3(specFile, "utf-8").split(`
`)[1]?.replace(/^>\s*/, "") || "" : "";
      console.log(`  ${source_default.green(k)}  ${source_default.gray(desc)}`);
    }
  }
  console.log(source_default.gray(`
  Community kits: npm install @scope/kit-<name> then dotdog kit install <name>`));
});
kitCmd.command("init <kit>").description("Init a project from a kit").option("-p, --project <name>").action((kit, opts) => {
  const src = join3(kitDir, kit);
  if (!existsSync3(src)) {
    console.log(source_default.red(`Kit "${kit}" not found. Available: ${builtInKits.join(", ")}`));
    return;
  }
  const projectName = opts.project || kit;
  const dir = resolvePath2(".");
  const dest = join3(dir, "specs", projectName);
  if (existsSync3(dest)) {
    console.log(source_default.yellow(`Project "${projectName}" already exists.`));
    return;
  }
  mkdirSync(dest, { recursive: true });
  const files = readdirSync3(src).filter((f) => f.endsWith(".dog"));
  for (const f of files) {
    writeFileSync2(join3(dest, f), readFileSync3(join3(src, f), "utf-8"));
    console.log(source_default.green(`  ✓ ${f}`));
  }
  console.log(source_default.gray(`
  Kit "${kit}" initialized in specs/${projectName}/`));
  console.log(source_default.gray(`  Run: dotdog validate`));
});
program2.command("badge [dir]").description("Generate dotdog-badge.svg (shields.io style) showing savings").action((d = ".") => {
  const dir = resolvePath2(d);
  const dirs = [join3(dir, "projects"), join3(dir, "specs"), dir];
  let found = false;
  for (const dd of dirs) {
    if (!existsSync3(dd))
      continue;
    const projects = readdirSync3(dd, { withFileTypes: true }).filter((e) => e.isDirectory()).map((e) => e.name);
    for (const p of projects) {
      const pd = join3(dd, p);
      const dagFile = join3(pd, `${p}.dag`);
      if (!existsSync3(join3(pd, "SPEC.dog")))
        continue;
      if (!existsSync3(dagFile)) {
        console.log(source_default.red(`  No .dag for ${p}. Run dotdog compile first.`));
        continue;
      }
      const rawDag = JSON.parse(readFileSync3(dagFile, "utf-8"));
      const dag = normalizeDag(rawDag);
      const saved = dag.tk && dag.tk.saved ? dag.tk.saved : 0;
      const fmt = saved >= 1000 ? `${(saved / 1000).toFixed(1)}K` : `${saved}`;
      const label = "dotdog";
      const value = `${fmt} tokens saved`;
      const pct = dag.tk && dag.tk.sv ? dag.tk.sv : 0;
      const color = pct > 90 ? "#4c1" : pct > 70 ? "#dfb317" : "#e05d44";
      const labelLen = Math.round(label.length * 7.2);
      const valueLen = Math.round(value.length * 7.2);
      const leftW = Math.max(labelLen + 10, 70);
      const rightW = Math.max(valueLen + 10, 40);
      const totalW = leftW + rightW;
      const leftX = leftW / 2 * 10;
      const rightX = (leftW + rightW / 2) * 10;
      const svg = `<svg xmlns="http://www.w3.org/2000/svg" width="${totalW}" height="20" role="img" aria-label="${label}: ${value}">
  <title>${label}: ${value}</title>
  <linearGradient id="s" x2="0" y2="100%"><stop offset="0" stop-color="#bbb" stop-opacity=".1"/><stop offset="1" stop-opacity=".1"/></linearGradient>
  <clipPath id="r"><rect width="${totalW}" height="20" rx="3" fill="#fff"/></clipPath>
  <g clip-path="url(#r)">
    <rect width="${leftW}" height="20" fill="#555"/>
    <rect x="${leftW}" width="${rightW}" height="20" fill="${color}"/>
    <rect width="${totalW}" height="20" fill="url(#s)"/>
  </g>
  <g fill="#fff" text-anchor="middle" font-family="Verdana,Geneva,DejaVu Sans,sans-serif" text-rendering="geometricPrecision" font-size="110">
    <g transform="scale(.1)">
      <text x="${leftX}" y="140" textLength="${labelLen * 10}">${label}</text>
      <text x="${rightX}" y="140" textLength="${valueLen * 10}">${value}</text>
    </g>
  </g>
</svg>`;
      writeFileSync2(join3(dir, "dotdog-badge.svg"), svg);
      console.log(source_default.green(`  ✓ dotdog-badge.svg  (${label}: ${value})`));
      found = true;
    }
  }
  if (!found)
    console.log(source_default.yellow("No projects found. Run dotdog init first."));
});
program2.command("convert <file>").description("Convert an .md file to .dog — rename + validate").action((f) => {
  const { existsSync: existsSync4, renameSync, readFileSync: readFileSync4, writeFileSync: writeFileSync3 } = __require("fs");
  const { join: join4, dirname, basename } = __require("path");
  const path2 = resolvePath2(f);
  if (!existsSync4(path2)) {
    console.log(source_default.red(`File not found: ${f}`));
    return;
  }
  if (!path2.endsWith(".md")) {
    console.log(source_default.yellow(`Only .md files can be converted.`));
    return;
  }
  const dogPath = path2.replace(/\.md$/, ".dog");
  if (existsSync4(dogPath)) {
    console.log(source_default.yellow(`${dogPath} already exists.`));
    return;
  }
  const content = readFileSync4(path2, "utf-8");
  renameSync(path2, dogPath);
  if (!content.match(/^##\s/m)) {
    const stub = `## Product

(Describe your product here)
`;
    writeFileSync3(dogPath, stub + content);
  }
  console.log(source_default.green(`  ✓ ${basename(path2)} → ${basename(dogPath)}`));
  console.log(source_default.gray("  Run: dotdog validate"));
});
program2.parse();
export {
  parseToJSON,
  parse
};
